\section{File List}
Here is a list of all documented files with brief descriptions\+:\begin{DoxyCompactList}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{1_88_84__epars_8h}{1.\+8.\+4\+\_\+epars.\+h} \\*Free energy parameters for parameter file conversion }{\pageref{1_88_84__epars_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{1_88_84__intloops_8h}{1.\+8.\+4\+\_\+intloops.\+h} \\*Free energy parameters for interior loop contributions needed by the parameter file conversion functions }{\pageref{1_88_84__intloops_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{2Dfold_8h}{2\+Dfold.\+h} }{\pageref{2Dfold_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{2Dpfold_8h}{2\+Dpfold.\+h} }{\pageref{2Dpfold_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/{\bfseries ali\+\_\+plex.\+h} }{\pageref{ali__plex_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{alifold_8h}{alifold.\+h} \\*Compute various properties (consensus M\+FE structures, partition function, Boltzmann distributed stochastic samples, ...) for R\+NA sequence alignments }{\pageref{alifold_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{aln__util_8h}{aln\+\_\+util.\+h} \\*Various utility-\/ and helper-\/functions for sequence alignments and comparative structure prediction }{\pageref{aln__util_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{alphabet_8h}{alphabet.\+h} \\*Functions to process, convert, and generally handle different nucleotide and/or base pair alphabets }{\pageref{alphabet_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{boltzmann__sampling_8h}{boltzmann\+\_\+sampling.\+h} \\*Boltzmann Sampling of secondary structures from the ensemble }{\pageref{boltzmann__sampling_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{centroid_8h}{centroid.\+h} \\*Centroid structure computation }{\pageref{centroid_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{cofold_8h}{cofold.\+h} \\*M\+FE version of cofolding routines }{\pageref{cofold_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{constraints_8h}{constraints.\+h} \\*Functions and data structures for constraining secondary structure predictions and evaluation }{\pageref{constraints_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{constraints__hard_8h}{constraints\+\_\+hard.\+h} \\*Functions and data structures for handling of secondary structure hard constraints }{\pageref{constraints__hard_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{constraints__SHAPE_8h}{constraints\+\_\+\+S\+H\+A\+P\+E.\+h} \\*This module provides function to incorporate S\+H\+A\+PE reactivity data into the folding recursions by means of soft constraints }{\pageref{constraints__SHAPE_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{constraints__soft_8h}{constraints\+\_\+soft.\+h} \\*Functions and data structures for secondary structure soft constraints }{\pageref{constraints__soft_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{convert__epars_8h}{convert\+\_\+epars.\+h} \\*Functions and definitions for energy parameter file format conversion }{\pageref{convert__epars_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{data__structures_8h}{data\+\_\+structures.\+h} }{\pageref{data__structures_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{dist__vars_8h}{dist\+\_\+vars.\+h} \\*Global variables for Distance-\/\+Package }{\pageref{dist__vars_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{dp__matrices_8h}{dp\+\_\+matrices.\+h} }{\pageref{dp__matrices_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{duplex_8h}{duplex.\+h} \\*Duplex folding function declarations.. }{\pageref{duplex_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{edit__cost_8h}{edit\+\_\+cost.\+h} \\*Global variables for Edit Costs included by treedist.\+c and stringdist.\+c }{\pageref{edit__cost_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{energy__const_8h}{energy\+\_\+const.\+h} }{\pageref{energy__const_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/{\bfseries energy\+\_\+par.\+h} }{\pageref{energy__par_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{equilibrium__probs_8h}{equilibrium\+\_\+probs.\+h} \\*Equilibrium Probability implementations }{\pageref{equilibrium__probs_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{eval_8h}{eval.\+h} \\*Functions and variables related to energy evaluation of sequence/structure pairs }{\pageref{eval_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{exterior__loops_8h}{exterior\+\_\+loops.\+h} \\*Energy evaluation of exterior loops for M\+FE and partition function calculations }{\pageref{exterior__loops_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{file__formats_8h}{file\+\_\+formats.\+h} \\*Functions dealing with file formats for R\+NA sequences, and structures }{\pageref{file__formats_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{file__formats__msa_8h}{file\+\_\+formats\+\_\+msa.\+h} \\*Functions dealing with file formats for Multiple Sequence Alignments (M\+SA) }{\pageref{file__formats__msa_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{findpath_8h}{findpath.\+h} }{\pageref{findpath_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{fold_8h}{fold.\+h} \\*M\+FE calculations for single R\+NA sequences }{\pageref{fold_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{fold__vars_8h}{fold\+\_\+vars.\+h} \\*Here all all declarations of the global variables used throughout R\+N\+Alib }{\pageref{fold__vars_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{gquad_8h}{gquad.\+h} \\*Various functions related to G-\/quadruplex computations }{\pageref{gquad_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{hairpin__loops_8h}{hairpin\+\_\+loops.\+h} \\*Energy evaluation of hairpin loops for M\+FE and partition function calculations }{\pageref{hairpin__loops_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{interior__loops_8h}{interior\+\_\+loops.\+h} \\*Energy evaluation of interior loops for M\+FE and partition function calculations }{\pageref{interior__loops_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/{\bfseries intl11.\+h} }{\pageref{intl11_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/{\bfseries intl11d\+H.\+h} }{\pageref{intl11dH_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/{\bfseries intl21.\+h} }{\pageref{intl21_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/{\bfseries intl21d\+H.\+h} }{\pageref{intl21dH_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/{\bfseries intl22.\+h} }{\pageref{intl22_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/{\bfseries intl22d\+H.\+h} }{\pageref{intl22dH_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{inverse_8h}{inverse.\+h} \\*Inverse folding routines }{\pageref{inverse_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{Lfold_8h}{Lfold.\+h} \\*Predicting local M\+FE structures of large sequences }{\pageref{Lfold_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{ligand_8h}{ligand.\+h} \\*Functions for incorporation of ligands binding to haipirn and interior loop motifs }{\pageref{ligand_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{loop__energies_8h}{loop\+\_\+energies.\+h} \\*Energy evaluation for M\+FE and partition function calculations }{\pageref{loop__energies_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{LPfold_8h}{L\+Pfold.\+h} \\*Function declarations of partition function variants of the Lfold algorithm }{\pageref{LPfold_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{MEA_8h}{M\+E\+A.\+h} \\*Computes a M\+EA (maximum expected accuracy) structure }{\pageref{MEA_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{mfe_8h}{mfe.\+h} \\*M\+FE calculations for single R\+NA sequences }{\pageref{mfe_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{mm_8h}{mm.\+h} \\*Several Maximum Matching implementations }{\pageref{mm_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{model_8h}{model.\+h} \\*The model details data structure and its corresponding modifiers }{\pageref{model_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/{\bfseries move\+\_\+set.\+h} }{\pageref{move__set_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{multibranch__loops_8h}{multibranch\+\_\+loops.\+h} \\*Energy evaluation of multibranch loops for M\+FE and partition function calculations }{\pageref{multibranch__loops_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{naview_8h}{naview.\+h} }{\pageref{naview_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/{\bfseries pair\+\_\+mat.\+h} }{\pageref{pair__mat_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{params_8h}{params.\+h} }{\pageref{params_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{part__func_8h}{part\+\_\+func.\+h} \\*Partition function of single R\+NA sequences }{\pageref{part__func_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{part__func__co_8h}{part\+\_\+func\+\_\+co.\+h} \\*Partition function for two R\+NA sequences }{\pageref{part__func__co_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{part__func__up_8h}{part\+\_\+func\+\_\+up.\+h} \\*Partition Function Cofolding as stepwise process }{\pageref{part__func__up_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{perturbation__fold_8h}{perturbation\+\_\+fold.\+h} \\*Find a vector of perturbation energies that minimizes the discripancies between predicted and observed pairing probabilities and the amount of neccessary adjustments }{\pageref{perturbation__fold_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/{\bfseries P\+Kplex.\+h} }{\pageref{PKplex_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/{\bfseries plex.\+h} }{\pageref{plex_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{plot__aln_8h}{plot\+\_\+aln.\+h} \\*Various functions for plotting Sequence / Structure Alignments }{\pageref{plot__aln_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{plot__layouts_8h}{plot\+\_\+layouts.\+h} \\*Secondary structure plot layout algorithms }{\pageref{plot__layouts_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{plot__structure_8h}{plot\+\_\+structure.\+h} \\*Various functions for plotting R\+NA secondary structures }{\pageref{plot__structure_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/{\bfseries Profile\+Aln.\+h} }{\pageref{ProfileAln_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{profiledist_8h}{profiledist.\+h} }{\pageref{profiledist_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{PS__dot_8h}{P\+S\+\_\+dot.\+h} \\*Various functions for plotting R\+NA secondary structures, dot-\/plots and other visualizations }{\pageref{PS__dot_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{read__epars_8h}{read\+\_\+epars.\+h} }{\pageref{read__epars_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{ribo_8h}{ribo.\+h} \\*Parse Ribo\+Sum Scoring Matrices for Covariance Scoring of Alignments }{\pageref{ribo_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{RNAstruct_8h}{R\+N\+Astruct.\+h} \\*Parsing and Coarse Graining of Structures }{\pageref{RNAstruct_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/{\bfseries snofold.\+h} }{\pageref{snofold_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/{\bfseries snoop.\+h} }{\pageref{snoop_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{string__utils_8h}{string\+\_\+utils.\+h} \\*General utility-\/ and helper-\/functions for R\+NA sequence and structure strings used throughout the Vienna\+R\+NA Package }{\pageref{string__utils_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{stringdist_8h}{stringdist.\+h} \\*Functions for String Alignment }{\pageref{stringdist_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{structure__utils_8h}{structure\+\_\+utils.\+h} \\*Various utility-\/ and helper-\/functions for secondary structure parsing, converting, etc }{\pageref{structure__utils_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{subopt_8h}{subopt.\+h} \\*R\+N\+Asubopt and density of states declarations }{\pageref{subopt_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/{\bfseries svm\+\_\+utils.\+h} }{\pageref{svm__utils_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{treedist_8h}{treedist.\+h} \\*Functions for \hyperlink{structTree}{Tree} Edit Distances }{\pageref{treedist_8h}}{}
\item\contentsline{section}{Vienna\+R\+N\+A/\hyperlink{utils_8h}{utils.\+h} \\*General utility-\/ and helper-\/functions used throughout the {\itshape Vienna\+R\+NA} {\itshape Package} }{\pageref{utils_8h}}{}
\end{DoxyCompactList}
