\hypertarget{group__file__utils}{}\section{Functions to Read/\+Write several File Formats for R\+NA Sequences, Structures, and Alignments}
\label{group__file__utils}\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
Collaboration diagram for Functions to Read/\+Write several File Formats for R\+NA Sequences, Structures, and Alignments\+:
\nopagebreak
\begin{figure}[H]
\begin{center}
\leavevmode
\includegraphics[width=327pt]{group__file__utils}
\end{center}
\end{figure}
\subsection*{Files}
\begin{DoxyCompactItemize}
\item 
file \hyperlink{file__formats_8h}{file\+\_\+formats.\+h}
\begin{DoxyCompactList}\small\item\em Functions dealing with file formats for R\+NA sequences, and structures. \end{DoxyCompactList}\item 
file \hyperlink{file__formats__msa_8h}{file\+\_\+formats\+\_\+msa.\+h}
\begin{DoxyCompactList}\small\item\em Functions dealing with file formats for Multiple Sequence Alignments (M\+SA) \end{DoxyCompactList}\item 
file \hyperlink{ribo_8h}{ribo.\+h}
\begin{DoxyCompactList}\small\item\em Parse Ribo\+Sum Scoring Matrices for Covariance Scoring of Alignments. \end{DoxyCompactList}\end{DoxyCompactItemize}
\subsection*{Macros}
\begin{DoxyCompactItemize}
\item 
\#define \hyperlink{group__file__utils_gabec89c09874528c6cb73140a4c3d86d7}{V\+R\+N\+A\+\_\+\+O\+P\+T\+I\+O\+N\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE}~32U
\begin{DoxyCompactList}\small\item\em Tell a function that an input is assumed to span several lines. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_ga7d725ef525b29891abef3f1ed42599a4}{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE}~32U
\begin{DoxyCompactList}\small\item\em parse multiline constraint \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_ga79a23de2c7249f2cccd762e475c81859}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL}~1U
\begin{DoxyCompactList}\small\item\em Option flag indicating ClustalW formatted files. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}~2U
\begin{DoxyCompactList}\small\item\em Option flag indicating Stockholm 1.\+0 formatted files. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_gacf6274a2c825f34e4131404665b00604}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA}~4U
\begin{DoxyCompactList}\small\item\em Option flag indicating F\+A\+S\+TA (Pearson) formatted files. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_gab3186d774c01570d2a47d0308eec4927}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF}~8U
\begin{DoxyCompactList}\small\item\em Option flag indicating M\+AF formatted files. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_ga4acc255373831856a8417b68de8a94c0}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}
\begin{DoxyCompactList}\small\item\em Option flag indicating the set of default file formats. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_ga229fb3778ecabea4782902b69fa48fd1}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK}~4096U
\begin{DoxyCompactList}\small\item\em Option flag to disable validation of the alignment. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_gabdc948f547e550125de3e7c65878400c}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN}~8192U
\begin{DoxyCompactList}\small\item\em Return flag of \hyperlink{group__file__utils_gade4fa8136ebb2d0f7eb3f8b59a8658e3}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} to indicate unknown or malformatted alignment. \end{DoxyCompactList}\end{DoxyCompactItemize}
\subsection*{Functions}
\begin{DoxyCompactItemize}
\item 
void \hyperlink{group__file__utils_gaaface7db12fadc3d271641c4515ab6e4}{vrna\+\_\+file\+\_\+helixlist} (const char $\ast$seq, const char $\ast$db, float energy, F\+I\+LE $\ast$file)
\begin{DoxyCompactList}\small\item\em Print a secondary structure as helix list. \end{DoxyCompactList}\item 
void \hyperlink{group__file__utils_gab69682373ccca1e0e28cc967eec07745}{vrna\+\_\+file\+\_\+connect} (const char $\ast$seq, const char $\ast$db, float energy, const char $\ast$identifier, F\+I\+LE $\ast$file)
\begin{DoxyCompactList}\small\item\em Print a secondary structure as connect table. \end{DoxyCompactList}\item 
void \hyperlink{group__file__utils_ga9b462e6f202594af5d3fa56e280d633f}{vrna\+\_\+file\+\_\+bpseq} (const char $\ast$seq, const char $\ast$db, F\+I\+LE $\ast$file)
\begin{DoxyCompactList}\small\item\em Print a secondary structure in bpseq format. \end{DoxyCompactList}\item 
void \hyperlink{group__file__utils_ga31f4a6c2ea1495a6e4f9eb45a9f6193d}{vrna\+\_\+file\+\_\+json} (const char $\ast$seq, const char $\ast$db, double energy, const char $\ast$identifier, F\+I\+LE $\ast$file)
\begin{DoxyCompactList}\small\item\em Print a secondary structure in jsonformat. \end{DoxyCompactList}\item 
unsigned int \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record} (char $\ast$$\ast$header, char $\ast$$\ast$sequence, char $\ast$$\ast$$\ast$rest, F\+I\+LE $\ast$file, unsigned int options)
\begin{DoxyCompactList}\small\item\em Get a (fasta) data set from a file or stdin. \end{DoxyCompactList}\item 
char $\ast$ \hyperlink{group__file__utils_gad37cbb63a05eed63ba25c91628409be0}{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure} (const char $\ast$$\ast$lines, unsigned int length, unsigned int option)
\begin{DoxyCompactList}\small\item\em Extract a dot-\/bracket structure string from (multiline)character array. \end{DoxyCompactList}\item 
int \hyperlink{group__file__utils_ga646ebf45450a69a7f2533f9ecd283a32}{vrna\+\_\+file\+\_\+\+S\+H\+A\+P\+E\+\_\+read} (const char $\ast$file\+\_\+name, int length, double default\+\_\+value, char $\ast$sequence, double $\ast$values)
\begin{DoxyCompactList}\small\item\em Read data from a given S\+H\+A\+PE reactivity input file. \end{DoxyCompactList}\item 
\hyperlink{group__data__structures_ga8e4eb5e1bfc95776559575beb359af87}{vrna\+\_\+plist\+\_\+t} $\ast$ \hyperlink{group__file__utils_gae33323c53765ecbbc410d9de2d495432}{vrna\+\_\+file\+\_\+constraints\+\_\+read} (const char $\ast$filename, unsigned int length, unsigned int options)
\begin{DoxyCompactList}\small\item\em Read constraints from an input file. \end{DoxyCompactList}\item 
void \hyperlink{group__file__utils_ga55a9ae6dfeecc1b3f0c2acf6fa796c15}{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+constraint} (char $\ast$$\ast$cstruc, const char $\ast$$\ast$lines, unsigned int option)
\begin{DoxyCompactList}\small\item\em Extract a hard constraint encoded as pseudo dot-\/bracket string. \end{DoxyCompactList}\item 
unsigned int \hyperlink{group__file__utils_gafd194a69af9d92b5b0412a7627ac1595}{read\+\_\+record} (char $\ast$$\ast$header, char $\ast$$\ast$sequence, char $\ast$$\ast$$\ast$rest, unsigned int options)
\begin{DoxyCompactList}\small\item\em Get a data record from stdin. \end{DoxyCompactList}\item 
int \hyperlink{group__file__utils_gad02d5d12bda54611c915a1019323b7be}{vrna\+\_\+file\+\_\+msa\+\_\+read} (const char $\ast$filename, char $\ast$$\ast$$\ast$names, char $\ast$$\ast$$\ast$aln, char $\ast$$\ast$id, char $\ast$$\ast$structure, unsigned int options)
\begin{DoxyCompactList}\small\item\em Read a multiple sequence alignment from file. \end{DoxyCompactList}\item 
int \hyperlink{group__file__utils_gad4203a438622b2df7bc2f16578d54799}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record} (F\+I\+LE $\ast$fp, char $\ast$$\ast$$\ast$names, char $\ast$$\ast$$\ast$aln, char $\ast$$\ast$id, char $\ast$$\ast$structure, unsigned int options)
\begin{DoxyCompactList}\small\item\em Read a multiple sequence alignment from file handle. \end{DoxyCompactList}\item 
unsigned int \hyperlink{group__file__utils_gade4fa8136ebb2d0f7eb3f8b59a8658e3}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format} (const char $\ast$filename, unsigned int options)
\begin{DoxyCompactList}\small\item\em Detect the format of a multiple sequence alignment file. \end{DoxyCompactList}\item 
float $\ast$$\ast$ \hyperlink{group__file__utils_ga5e125c9586fcd4e2e1559fe76f7289cc}{readribosum} (char $\ast$name)\hypertarget{group__file__utils_ga5e125c9586fcd4e2e1559fe76f7289cc}{}\label{group__file__utils_ga5e125c9586fcd4e2e1559fe76f7289cc}

\begin{DoxyCompactList}\small\item\em Read a Ribo\+Sum or other user-\/defined Scoring Matrix and Store into global Memory. \end{DoxyCompactList}\end{DoxyCompactItemize}


\subsection{Detailed Description}


\subsection{Macro Definition Documentation}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+O\+P\+T\+I\+O\+N\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE@{V\+R\+N\+A\+\_\+\+O\+P\+T\+I\+O\+N\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE}}
\index{V\+R\+N\+A\+\_\+\+O\+P\+T\+I\+O\+N\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE@{V\+R\+N\+A\+\_\+\+O\+P\+T\+I\+O\+N\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{V\+R\+N\+A\+\_\+\+O\+P\+T\+I\+O\+N\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE}{VRNA_OPTION_MULTILINE}}]{\setlength{\rightskip}{0pt plus 5cm}\#define V\+R\+N\+A\+\_\+\+O\+P\+T\+I\+O\+N\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE~32U}\hypertarget{group__file__utils_gabec89c09874528c6cb73140a4c3d86d7}{}\label{group__file__utils_gabec89c09874528c6cb73140a4c3d86d7}


{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Tell a function that an input is assumed to span several lines. 

If used as input-\/option a function might also be returning this state telling that it has read data from multiple lines.

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_gad37cbb63a05eed63ba25c91628409be0}{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure()}, \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()} 
\end{DoxySeeAlso}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE@{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE}}
\index{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE@{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE}{VRNA_CONSTRAINT_MULTILINE}}]{\setlength{\rightskip}{0pt plus 5cm}\#define V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE~32U}\hypertarget{group__file__utils_ga7d725ef525b29891abef3f1ed42599a4}{}\label{group__file__utils_ga7d725ef525b29891abef3f1ed42599a4}


{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



parse multiline constraint 

\begin{DoxyRefDesc}{Deprecated}
\item[\hyperlink{deprecated__deprecated000065}{Deprecated}]see \hyperlink{group__file__utils_gad37cbb63a05eed63ba25c91628409be0}{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure()} \end{DoxyRefDesc}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL}{VRNA_FILE_FORMAT_MSA_CLUSTAL}}]{\setlength{\rightskip}{0pt plus 5cm}\#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL~1U}\hypertarget{group__file__utils_ga79a23de2c7249f2cccd762e475c81859}{}\label{group__file__utils_ga79a23de2c7249f2cccd762e475c81859}


{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Option flag indicating ClustalW formatted files. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_gad02d5d12bda54611c915a1019323b7be}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, \hyperlink{group__file__utils_gad4203a438622b2df7bc2f16578d54799}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()}, \hyperlink{group__file__utils_gade4fa8136ebb2d0f7eb3f8b59a8658e3}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} 
\end{DoxySeeAlso}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}{VRNA_FILE_FORMAT_MSA_STOCKHOLM}}]{\setlength{\rightskip}{0pt plus 5cm}\#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM~2U}\hypertarget{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}{}\label{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}


{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Option flag indicating Stockholm 1.\+0 formatted files. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_gad02d5d12bda54611c915a1019323b7be}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, \hyperlink{group__file__utils_gad4203a438622b2df7bc2f16578d54799}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()}, \hyperlink{group__file__utils_gade4fa8136ebb2d0f7eb3f8b59a8658e3}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} 
\end{DoxySeeAlso}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA}{VRNA_FILE_FORMAT_MSA_FASTA}}]{\setlength{\rightskip}{0pt plus 5cm}\#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA~4U}\hypertarget{group__file__utils_gacf6274a2c825f34e4131404665b00604}{}\label{group__file__utils_gacf6274a2c825f34e4131404665b00604}


{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Option flag indicating F\+A\+S\+TA (Pearson) formatted files. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_gad02d5d12bda54611c915a1019323b7be}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, \hyperlink{group__file__utils_gad4203a438622b2df7bc2f16578d54799}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()}, \hyperlink{group__file__utils_gade4fa8136ebb2d0f7eb3f8b59a8658e3}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} 
\end{DoxySeeAlso}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF}{VRNA_FILE_FORMAT_MSA_MAF}}]{\setlength{\rightskip}{0pt plus 5cm}\#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF~8U}\hypertarget{group__file__utils_gab3186d774c01570d2a47d0308eec4927}{}\label{group__file__utils_gab3186d774c01570d2a47d0308eec4927}


{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Option flag indicating M\+AF formatted files. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_gad02d5d12bda54611c915a1019323b7be}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, \hyperlink{group__file__utils_gad4203a438622b2df7bc2f16578d54799}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()}, \hyperlink{group__file__utils_gade4fa8136ebb2d0f7eb3f8b59a8658e3}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} 
\end{DoxySeeAlso}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}{VRNA_FILE_FORMAT_MSA_DEFAULT}}]{\setlength{\rightskip}{0pt plus 5cm}\#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}\hypertarget{group__file__utils_ga4acc255373831856a8417b68de8a94c0}{}\label{group__file__utils_ga4acc255373831856a8417b68de8a94c0}


{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}

{\bfseries Value\+:}
\begin{DoxyCode}
( \(\backslash\)
                                              VRNA\_FILE\_FORMAT\_MSA\_CLUSTAL \(\backslash\)
                                            | \hyperlink{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}{VRNA\_FILE\_FORMAT\_MSA\_STOCKHOLM} 
      \(\backslash\)
                                            | \hyperlink{group__file__utils_gacf6274a2c825f34e4131404665b00604}{VRNA\_FILE\_FORMAT\_MSA\_FASTA} \(\backslash\)
                                            | \hyperlink{group__file__utils_gab3186d774c01570d2a47d0308eec4927}{VRNA\_FILE\_FORMAT\_MSA\_MAF} \(\backslash\)
                                          )
\end{DoxyCode}


Option flag indicating the set of default file formats. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_gad02d5d12bda54611c915a1019323b7be}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, \hyperlink{group__file__utils_gad4203a438622b2df7bc2f16578d54799}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()}, \hyperlink{group__file__utils_gade4fa8136ebb2d0f7eb3f8b59a8658e3}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} 
\end{DoxySeeAlso}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK}{VRNA_FILE_FORMAT_MSA_NOCHECK}}]{\setlength{\rightskip}{0pt plus 5cm}\#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK~4096U}\hypertarget{group__file__utils_ga229fb3778ecabea4782902b69fa48fd1}{}\label{group__file__utils_ga229fb3778ecabea4782902b69fa48fd1}


{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Option flag to disable validation of the alignment. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_gad02d5d12bda54611c915a1019323b7be}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, \hyperlink{group__file__utils_gad4203a438622b2df7bc2f16578d54799}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()} 
\end{DoxySeeAlso}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN}{VRNA_FILE_FORMAT_MSA_UNKNOWN}}]{\setlength{\rightskip}{0pt plus 5cm}\#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN~8192U}\hypertarget{group__file__utils_gabdc948f547e550125de3e7c65878400c}{}\label{group__file__utils_gabdc948f547e550125de3e7c65878400c}


{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Return flag of \hyperlink{group__file__utils_gade4fa8136ebb2d0f7eb3f8b59a8658e3}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} to indicate unknown or malformatted alignment. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_gade4fa8136ebb2d0f7eb3f8b59a8658e3}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} 
\end{DoxySeeAlso}


\subsection{Function Documentation}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+helixlist@{vrna\+\_\+file\+\_\+helixlist}}
\index{vrna\+\_\+file\+\_\+helixlist@{vrna\+\_\+file\+\_\+helixlist}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{vrna\+\_\+file\+\_\+helixlist(const char $\ast$seq, const char $\ast$db, float energy, F\+I\+L\+E $\ast$file)}{vrna_file_helixlist(const char *seq, const char *db, float energy, FILE *file)}}]{\setlength{\rightskip}{0pt plus 5cm}void vrna\+\_\+file\+\_\+helixlist (
\begin{DoxyParamCaption}
\item[{const char $\ast$}]{seq, }
\item[{const char $\ast$}]{db, }
\item[{float}]{energy, }
\item[{F\+I\+LE $\ast$}]{file}
\end{DoxyParamCaption}
)}\hypertarget{group__file__utils_gaaface7db12fadc3d271641c4515ab6e4}{}\label{group__file__utils_gaaface7db12fadc3d271641c4515ab6e4}


{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Print a secondary structure as helix list. 


\begin{DoxyParams}{Parameters}
{\em seq} & The R\+NA sequence \\
\hline
{\em db} & The structure in dot-\/bracket format \\
\hline
{\em energy} & Free energy of the structure in kcal/mol \\
\hline
{\em file} & The file handle used to print to (print defaults to \textquotesingle{}stdout\textquotesingle{} if(file == N\+U\+LL) ) \\
\hline
\end{DoxyParams}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+connect@{vrna\+\_\+file\+\_\+connect}}
\index{vrna\+\_\+file\+\_\+connect@{vrna\+\_\+file\+\_\+connect}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{vrna\+\_\+file\+\_\+connect(const char $\ast$seq, const char $\ast$db, float energy, const char $\ast$identifier, F\+I\+L\+E $\ast$file)}{vrna_file_connect(const char *seq, const char *db, float energy, const char *identifier, FILE *file)}}]{\setlength{\rightskip}{0pt plus 5cm}void vrna\+\_\+file\+\_\+connect (
\begin{DoxyParamCaption}
\item[{const char $\ast$}]{seq, }
\item[{const char $\ast$}]{db, }
\item[{float}]{energy, }
\item[{const char $\ast$}]{identifier, }
\item[{F\+I\+LE $\ast$}]{file}
\end{DoxyParamCaption}
)}\hypertarget{group__file__utils_gab69682373ccca1e0e28cc967eec07745}{}\label{group__file__utils_gab69682373ccca1e0e28cc967eec07745}


{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Print a secondary structure as connect table. 

Connect table file format looks like this\+: \begin{DoxyVerb}300  ENERGY = 7.0  example
  1 G       0    2   22    1
  2 G       1    3   21    2
\end{DoxyVerb}
 where the headerline is followed by 6 columns with\+:
\begin{DoxyEnumerate}
\item Base number\+: index n
\item Base (A, C, G, T, U, X)
\item Index n-\/1 (0 if first nucleotide)
\item Index n+1 (0 if last nucleotide)
\item Number of the base to which n is paired. No pairing is indicated by 0 (zero).
\item Natural numbering.
\end{DoxyEnumerate}


\begin{DoxyParams}{Parameters}
{\em seq} & The R\+NA sequence \\
\hline
{\em db} & The structure in dot-\/bracket format \\
\hline
{\em energy} & The free energy of the structure \\
\hline
{\em identifier} & An optional identifier for the sequence \\
\hline
{\em file} & The file handle used to print to (print defaults to \textquotesingle{}stdout\textquotesingle{} if(file == N\+U\+LL) ) \\
\hline
\end{DoxyParams}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+bpseq@{vrna\+\_\+file\+\_\+bpseq}}
\index{vrna\+\_\+file\+\_\+bpseq@{vrna\+\_\+file\+\_\+bpseq}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{vrna\+\_\+file\+\_\+bpseq(const char $\ast$seq, const char $\ast$db, F\+I\+L\+E $\ast$file)}{vrna_file_bpseq(const char *seq, const char *db, FILE *file)}}]{\setlength{\rightskip}{0pt plus 5cm}void vrna\+\_\+file\+\_\+bpseq (
\begin{DoxyParamCaption}
\item[{const char $\ast$}]{seq, }
\item[{const char $\ast$}]{db, }
\item[{F\+I\+LE $\ast$}]{file}
\end{DoxyParamCaption}
)}\hypertarget{group__file__utils_ga9b462e6f202594af5d3fa56e280d633f}{}\label{group__file__utils_ga9b462e6f202594af5d3fa56e280d633f}


{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Print a secondary structure in bpseq format. 


\begin{DoxyParams}{Parameters}
{\em seq} & The R\+NA sequence \\
\hline
{\em db} & The structure in dot-\/bracket format \\
\hline
{\em file} & The file handle used to print to (print defaults to \textquotesingle{}stdout\textquotesingle{} if(file == N\+U\+LL) ) \\
\hline
\end{DoxyParams}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+json@{vrna\+\_\+file\+\_\+json}}
\index{vrna\+\_\+file\+\_\+json@{vrna\+\_\+file\+\_\+json}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{vrna\+\_\+file\+\_\+json(const char $\ast$seq, const char $\ast$db, double energy, const char $\ast$identifier, F\+I\+L\+E $\ast$file)}{vrna_file_json(const char *seq, const char *db, double energy, const char *identifier, FILE *file)}}]{\setlength{\rightskip}{0pt plus 5cm}void vrna\+\_\+file\+\_\+json (
\begin{DoxyParamCaption}
\item[{const char $\ast$}]{seq, }
\item[{const char $\ast$}]{db, }
\item[{double}]{energy, }
\item[{const char $\ast$}]{identifier, }
\item[{F\+I\+LE $\ast$}]{file}
\end{DoxyParamCaption}
)}\hypertarget{group__file__utils_ga31f4a6c2ea1495a6e4f9eb45a9f6193d}{}\label{group__file__utils_ga31f4a6c2ea1495a6e4f9eb45a9f6193d}


{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Print a secondary structure in jsonformat. 


\begin{DoxyParams}{Parameters}
{\em seq} & The R\+NA sequence \\
\hline
{\em db} & The structure in dot-\/bracket format \\
\hline
{\em energy} & The free energy \\
\hline
{\em identifier} & An identifier for the sequence \\
\hline
{\em file} & The file handle used to print to (print defaults to \textquotesingle{}stdout\textquotesingle{} if(file == N\+U\+LL) ) \\
\hline
\end{DoxyParams}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record@{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record}}
\index{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record@{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record(char $\ast$$\ast$header, char $\ast$$\ast$sequence, char $\ast$$\ast$$\ast$rest, F\+I\+L\+E $\ast$file, unsigned int options)}{vrna_file_fasta_read_record(char **header, char **sequence, char ***rest, FILE *file, unsigned int options)}}]{\setlength{\rightskip}{0pt plus 5cm}unsigned int vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record (
\begin{DoxyParamCaption}
\item[{char $\ast$$\ast$}]{header, }
\item[{char $\ast$$\ast$}]{sequence, }
\item[{char $\ast$$\ast$$\ast$}]{rest, }
\item[{F\+I\+LE $\ast$}]{file, }
\item[{unsigned int}]{options}
\end{DoxyParamCaption}
)}\hypertarget{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{}\label{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}


{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Get a (fasta) data set from a file or stdin. 

This function may be used to obtain complete datasets from a filehandle or stdin. A dataset is always defined to contain at least a sequence. If data starts with a fasta header, i.\+e. a line like \begin{DoxyVerb}>some header info \end{DoxyVerb}
 then \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()} will assume that the sequence that follows the header may span over several lines. To disable this behavior and to assign a single line to the argument \textquotesingle{}sequence\textquotesingle{} one can pass \hyperlink{group__utils_ga0de536599b881c787b0943a2671da476}{V\+R\+N\+A\+\_\+\+I\+N\+P\+U\+T\+\_\+\+N\+O\+\_\+\+S\+P\+AN} in the \textquotesingle{}options\textquotesingle{} argument. If no fasta header is read in the beginning of a data block, a sequence must not span over multiple lines!~\newline
 Unless the options \hyperlink{group__utils_ga0f6311f11bed1842e3a527ab27b294c6}{V\+R\+N\+A\+\_\+\+I\+N\+P\+U\+T\+\_\+\+N\+O\+S\+K\+I\+P\+\_\+\+C\+O\+M\+M\+E\+N\+TS} or \hyperlink{group__utils_gab4db885222b3b69608310d7c7e63e286}{V\+R\+N\+A\+\_\+\+I\+N\+P\+U\+T\+\_\+\+N\+O\+S\+K\+I\+P\+\_\+\+B\+L\+A\+N\+K\+\_\+\+L\+I\+N\+ES} are passed, a sequence may be interrupted by lines starting with a comment character or empty lines.~\newline
 A sequence is regarded as completely read if it was either assumed to not span over multiple lines, a secondary structure or structure constraint follows the sequence on the next line, or a new header marks the beginning of a new sequence...~\newline
 All lines following the sequence (this includes comments) that do not initiate a new dataset according to the above definition are available through the line-\/array \textquotesingle{}rest\textquotesingle{}. Here one can usually find the structure constraint or other information belonging to the current dataset. Filling of \textquotesingle{}rest\textquotesingle{} may be prevented by passing \hyperlink{group__utils_ga7a2e8c50a0c7ce82e60da1016e1367fd}{V\+R\+N\+A\+\_\+\+I\+N\+P\+U\+T\+\_\+\+N\+O\+\_\+\+R\+E\+ST} to the options argument.~\newline
 \begin{DoxyNote}{Note}
This function will exit any program with an error message if no sequence could be read! 

This function is N\+OT threadsafe! It uses a global variable to store information about the next data block.
\end{DoxyNote}
The main purpose of this function is to be able to easily parse blocks of data in the header of a loop where all calculations for the appropriate data is done inside the loop. The loop may be then left on certain return values, e.\+g.\+: 
\begin{DoxyCode}
00001 char *id, *seq, **rest;
00002 int  i;
00003 id = seq = NULL;
00004 rest = NULL;
00005 while(!(vrna\_file\_fasta\_read\_record(&id, &seq, &rest, NULL, 0) & (VRNA\_INPUT\_ERROR |
       VRNA\_INPUT\_QUIT)))\{
00006   if(id) printf("%s\(\backslash\)n", id);
00007   printf("%s\(\backslash\)n", seq);
00008   if(rest)
00009     for(i=0;rest[i];i++)\{
00010       printf("%s\(\backslash\)n", rest[i]);
00011       free(rest[i]);
00012     \}
00013   free(rest);
00014   free(seq);
00015   free(id);
00016 \}
\end{DoxyCode}
 In the example above, the while loop will be terminated when \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()} returns either an error, E\+OF, or a user initiated quit request.~\newline
 As long as data is read from stdin (we are passing N\+U\+LL as the file pointer), the id is printed if it is available for the current block of data. The sequence will be printed in any case and if some more lines belong to the current block of data each line will be printed as well.

\begin{DoxyNote}{Note}
Do not forget to free the memory occupied by header, sequence and rest!
\end{DoxyNote}

\begin{DoxyParams}{Parameters}
{\em header} & A pointer which will be set such that it points to the header of the record \\
\hline
{\em sequence} & A pointer which will be set such that it points to the sequence of the record \\
\hline
{\em rest} & A pointer which will be set such that it points to an array of lines which also belong to the record \\
\hline
{\em file} & A file handle to read from (if N\+U\+LL, this function reads from stdin) \\
\hline
{\em options} & Some options which may be passed to alter the behavior of the function, use 0 for no options \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
A flag with information about what the function actually did read 
\end{DoxyReturn}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure@{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure}}
\index{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure@{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure(const char $\ast$$\ast$lines, unsigned int length, unsigned int option)}{vrna_extract_record_rest_structure(const char **lines, unsigned int length, unsigned int option)}}]{\setlength{\rightskip}{0pt plus 5cm}char$\ast$ vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure (
\begin{DoxyParamCaption}
\item[{const char $\ast$$\ast$}]{lines, }
\item[{unsigned int}]{length, }
\item[{unsigned int}]{option}
\end{DoxyParamCaption}
)}\hypertarget{group__file__utils_gad37cbb63a05eed63ba25c91628409be0}{}\label{group__file__utils_gad37cbb63a05eed63ba25c91628409be0}


{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Extract a dot-\/bracket structure string from (multiline)character array. 

This function extracts a dot-\/bracket structure string from the \textquotesingle{}rest\textquotesingle{} array as returned by \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()} and returns it. All occurences of comments within the \textquotesingle{}lines\textquotesingle{} array will be skipped as long as they do not break the structure string. If no structure could be read, this function returns N\+U\+LL.

\begin{DoxyPrecond}{Precondition}
The argument \textquotesingle{}lines\textquotesingle{} has to be a 2-\/dimensional character array as obtained by \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()} 
\end{DoxyPrecond}
\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()}
\end{DoxySeeAlso}

\begin{DoxyParams}{Parameters}
{\em lines} & The (multiline) character array to be parsed \\
\hline
{\em length} & The assumed length of the dot-\/bracket string (passing a value $<$ 1 results in no length limit) \\
\hline
{\em option} & Some options which may be passed to alter the behavior of the function, use 0 for no options \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
The dot-\/bracket string read from lines or N\+U\+LL 
\end{DoxyReturn}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+\+S\+H\+A\+P\+E\+\_\+read@{vrna\+\_\+file\+\_\+\+S\+H\+A\+P\+E\+\_\+read}}
\index{vrna\+\_\+file\+\_\+\+S\+H\+A\+P\+E\+\_\+read@{vrna\+\_\+file\+\_\+\+S\+H\+A\+P\+E\+\_\+read}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{vrna\+\_\+file\+\_\+\+S\+H\+A\+P\+E\+\_\+read(const char $\ast$file\+\_\+name, int length, double default\+\_\+value, char $\ast$sequence, double $\ast$values)}{vrna_file_SHAPE_read(const char *file_name, int length, double default_value, char *sequence, double *values)}}]{\setlength{\rightskip}{0pt plus 5cm}int vrna\+\_\+file\+\_\+\+S\+H\+A\+P\+E\+\_\+read (
\begin{DoxyParamCaption}
\item[{const char $\ast$}]{file\+\_\+name, }
\item[{int}]{length, }
\item[{double}]{default\+\_\+value, }
\item[{char $\ast$}]{sequence, }
\item[{double $\ast$}]{values}
\end{DoxyParamCaption}
)}\hypertarget{group__file__utils_ga646ebf45450a69a7f2533f9ecd283a32}{}\label{group__file__utils_ga646ebf45450a69a7f2533f9ecd283a32}


{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Read data from a given S\+H\+A\+PE reactivity input file. 

This function parses the informations from a given file and stores the result in the preallocated string sequence and the double array values.


\begin{DoxyParams}{Parameters}
{\em file\+\_\+name} & Path to the constraints file \\
\hline
{\em length} & Length of the sequence (file entries exceeding this limit will cause an error) \\
\hline
{\em default\+\_\+value} & Value for missing indices \\
\hline
{\em sequence} & Pointer to an array used for storing the sequence obtained from the S\+H\+A\+PE reactivity file \\
\hline
{\em values} & Pointer to an array used for storing the values obtained from the S\+H\+A\+PE reactivity file \\
\hline
\end{DoxyParams}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+constraints\+\_\+read@{vrna\+\_\+file\+\_\+constraints\+\_\+read}}
\index{vrna\+\_\+file\+\_\+constraints\+\_\+read@{vrna\+\_\+file\+\_\+constraints\+\_\+read}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{vrna\+\_\+file\+\_\+constraints\+\_\+read(const char $\ast$filename, unsigned int length, unsigned int options)}{vrna_file_constraints_read(const char *filename, unsigned int length, unsigned int options)}}]{\setlength{\rightskip}{0pt plus 5cm}{\bf vrna\+\_\+plist\+\_\+t}$\ast$ vrna\+\_\+file\+\_\+constraints\+\_\+read (
\begin{DoxyParamCaption}
\item[{const char $\ast$}]{filename, }
\item[{unsigned int}]{length, }
\item[{unsigned int}]{options}
\end{DoxyParamCaption}
)}\hypertarget{group__file__utils_gae33323c53765ecbbc410d9de2d495432}{}\label{group__file__utils_gae33323c53765ecbbc410d9de2d495432}


{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Read constraints from an input file. 

This function reads constraint definitions from a file and converts them into an array of \hyperlink{group__data__structures_ga8e4eb5e1bfc95776559575beb359af87}{vrna\+\_\+plist\+\_\+t} data structures. The data fields of each individual returned plist entry may adopt the following configurations\+:
\begin{DoxyItemize}
\item plist.\+i == plist.\+j $ \rightarrow $ single nucleotide constraint
\item plist.\+i != plist.\+j $ \rightarrow $ base pair constraint
\item plist.\+i == 0 $ \rightarrow $ End of list 
\end{DoxyItemize}\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+constraint@{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+constraint}}
\index{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+constraint@{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+constraint}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+constraint(char $\ast$$\ast$cstruc, const char $\ast$$\ast$lines, unsigned int option)}{vrna_extract_record_rest_constraint(char **cstruc, const char **lines, unsigned int option)}}]{\setlength{\rightskip}{0pt plus 5cm}void vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+constraint (
\begin{DoxyParamCaption}
\item[{char $\ast$$\ast$}]{cstruc, }
\item[{const char $\ast$$\ast$}]{lines, }
\item[{unsigned int}]{option}
\end{DoxyParamCaption}
)}\hypertarget{group__file__utils_ga55a9ae6dfeecc1b3f0c2acf6fa796c15}{}\label{group__file__utils_ga55a9ae6dfeecc1b3f0c2acf6fa796c15}


{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Extract a hard constraint encoded as pseudo dot-\/bracket string. 

\begin{DoxyRefDesc}{Deprecated}
\item[\hyperlink{deprecated__deprecated000066}{Deprecated}]Use \hyperlink{group__file__utils_gad37cbb63a05eed63ba25c91628409be0}{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure()} instead! \end{DoxyRefDesc}
\begin{DoxyPrecond}{Precondition}
The argument \textquotesingle{}lines\textquotesingle{} has to be a 2-\/dimensional character array as obtained by \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()} 
\end{DoxyPrecond}
\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()}, \hyperlink{group__hard__constraints_ga13053547a2de5532b64b64d35e097ae1}{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+D\+B\+\_\+\+P\+I\+PE}, \hyperlink{group__hard__constraints_ga369bea82eae75fbe626f409fa425747e}{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+D\+B\+\_\+\+D\+OT}, \hyperlink{group__hard__constraints_ga7283bbe0f8954f7b030ecc3f2d1932b2}{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+D\+B\+\_\+X} \hyperlink{constraints__hard_8h_ad54c1315a47d55653dcaa5de6e544b77}{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+D\+B\+\_\+\+A\+N\+G\+\_\+\+B\+R\+A\+CK}, \hyperlink{group__hard__constraints_gac17b034852c914bc5879954c65d7e74b}{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+D\+B\+\_\+\+R\+N\+D\+\_\+\+B\+R\+A\+CK}
\end{DoxySeeAlso}

\begin{DoxyParams}{Parameters}
{\em cstruc} & A pointer to a character array that is used as pseudo dot-\/bracket output \\
\hline
{\em lines} & A 2-\/dimensional character array with the extension lines from the F\+A\+S\+TA input \\
\hline
{\em option} & The option flags that define the behavior and recognition pattern of this function \\
\hline
\end{DoxyParams}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!read\+\_\+record@{read\+\_\+record}}
\index{read\+\_\+record@{read\+\_\+record}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{read\+\_\+record(char $\ast$$\ast$header, char $\ast$$\ast$sequence, char $\ast$$\ast$$\ast$rest, unsigned int options)}{read_record(char **header, char **sequence, char ***rest, unsigned int options)}}]{\setlength{\rightskip}{0pt plus 5cm}unsigned int read\+\_\+record (
\begin{DoxyParamCaption}
\item[{char $\ast$$\ast$}]{header, }
\item[{char $\ast$$\ast$}]{sequence, }
\item[{char $\ast$$\ast$$\ast$}]{rest, }
\item[{unsigned int}]{options}
\end{DoxyParamCaption}
)}\hypertarget{group__file__utils_gafd194a69af9d92b5b0412a7627ac1595}{}\label{group__file__utils_gafd194a69af9d92b5b0412a7627ac1595}


{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Get a data record from stdin. 

\begin{DoxyRefDesc}{Deprecated}
\item[\hyperlink{deprecated__deprecated000068}{Deprecated}]This function is deprecated! Use \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()} as a replacment.\end{DoxyRefDesc}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+msa\+\_\+read@{vrna\+\_\+file\+\_\+msa\+\_\+read}}
\index{vrna\+\_\+file\+\_\+msa\+\_\+read@{vrna\+\_\+file\+\_\+msa\+\_\+read}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{vrna\+\_\+file\+\_\+msa\+\_\+read(const char $\ast$filename, char $\ast$$\ast$$\ast$names, char $\ast$$\ast$$\ast$aln, char $\ast$$\ast$id, char $\ast$$\ast$structure, unsigned int options)}{vrna_file_msa_read(const char *filename, char ***names, char ***aln, char **id, char **structure, unsigned int options)}}]{\setlength{\rightskip}{0pt plus 5cm}int vrna\+\_\+file\+\_\+msa\+\_\+read (
\begin{DoxyParamCaption}
\item[{const char $\ast$}]{filename, }
\item[{char $\ast$$\ast$$\ast$}]{names, }
\item[{char $\ast$$\ast$$\ast$}]{aln, }
\item[{char $\ast$$\ast$}]{id, }
\item[{char $\ast$$\ast$}]{structure, }
\item[{unsigned int}]{options}
\end{DoxyParamCaption}
)}\hypertarget{group__file__utils_gad02d5d12bda54611c915a1019323b7be}{}\label{group__file__utils_gad02d5d12bda54611c915a1019323b7be}


{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Read a multiple sequence alignment from file. 

This function reads the (first) multiple sequence alignment from an input file. The read alignment is split into the sequence id/name part and the actual sequence information and stored in memory as arrays of ids/names and sequences. If the alignment file format allows for additional information, such as an ID of the entire alignment or consensus structure information, this data is retrieved as well and made available. The {\ttfamily options} parameter allows to specify the set of alignment file formats that should be used to retrieve the data. If 0 is passed as option, the list of alignment file formats defaults to \hyperlink{group__file__utils_ga4acc255373831856a8417b68de8a94c0}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}.

Currently, the list of parsable multiple sequence alignment file formats consists of\+:
\begin{DoxyItemize}
\item \hyperlink{file_formats_msa-formats-clustal}{ClustalW format}
\item \hyperlink{file_formats_msa-formats-stockholm}{Stockholm 1.\+0 format}
\item \hyperlink{file_formats_msa-formats-fasta}{F\+A\+S\+TA (Pearson) format}
\item \hyperlink{file_formats_msa-formats-maf}{M\+AF format}
\end{DoxyItemize}\begin{DoxyNote}{Note}
After successfully reading an alignment, this function performs a validation of the data that includes uniqueness of the sequence identifiers, and equal sequence lengths. This check can be deactivated by passing \hyperlink{group__file__utils_ga229fb3778ecabea4782902b69fa48fd1}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK} in the {\ttfamily options} parameter.
\end{DoxyNote}
\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_gad4203a438622b2df7bc2f16578d54799}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()}, \hyperlink{group__file__utils_ga79a23de2c7249f2cccd762e475c81859}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL}, \hyperlink{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}, \hyperlink{group__file__utils_gacf6274a2c825f34e4131404665b00604}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA}, \hyperlink{group__file__utils_gab3186d774c01570d2a47d0308eec4927}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF}, \hyperlink{group__file__utils_ga4acc255373831856a8417b68de8a94c0}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}, \hyperlink{group__file__utils_ga229fb3778ecabea4782902b69fa48fd1}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK}
\end{DoxySeeAlso}

\begin{DoxyParams}{Parameters}
{\em filename} & The name of input file that contains the alignment \\
\hline
{\em names} & An address to the pointer where sequence identifiers should be written to \\
\hline
{\em aln} & An address to the pointer where aligned sequences should be written to \\
\hline
{\em id} & An address to the pointer where the alignment ID should be written to (Maybe N\+U\+LL) \\
\hline
{\em structure} & An address to the pointer where consensus structure information should be written to (Maybe N\+U\+LL) \\
\hline
{\em options} & Options to manipulate the behavior of this function \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
The number of sequences in the alignment, or -\/1 if no alignment record could be found 
\end{DoxyReturn}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record@{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record}}
\index{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record@{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record(\+F\+I\+L\+E $\ast$fp, char $\ast$$\ast$$\ast$names, char $\ast$$\ast$$\ast$aln, char $\ast$$\ast$id, char $\ast$$\ast$structure, unsigned int options)}{vrna_file_msa_read_record(FILE *fp, char ***names, char ***aln, char **id, char **structure, unsigned int options)}}]{\setlength{\rightskip}{0pt plus 5cm}int vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record (
\begin{DoxyParamCaption}
\item[{F\+I\+LE $\ast$}]{fp, }
\item[{char $\ast$$\ast$$\ast$}]{names, }
\item[{char $\ast$$\ast$$\ast$}]{aln, }
\item[{char $\ast$$\ast$}]{id, }
\item[{char $\ast$$\ast$}]{structure, }
\item[{unsigned int}]{options}
\end{DoxyParamCaption}
)}\hypertarget{group__file__utils_gad4203a438622b2df7bc2f16578d54799}{}\label{group__file__utils_gad4203a438622b2df7bc2f16578d54799}


{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Read a multiple sequence alignment from file handle. 

Similar to \hyperlink{group__file__utils_gad02d5d12bda54611c915a1019323b7be}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, this function reads a multiple sequence alignment from an input file handle. Since using a file handle, this function is not limited to the first alignment record, but allows for looping over all alignments within the input.

The read alignment is split into the sequence id/name part and the actual sequence information and stored in memory as arrays of ids/names and sequences. If the alignment file format allows for additional information, such as an ID of the entire alignment or consensus structure information, this data is retrieved as well and made available. The {\ttfamily options} parameter allows to specify the alignment file format used to retrieve the data. A single format must be specified here, see \hyperlink{group__file__utils_gade4fa8136ebb2d0f7eb3f8b59a8658e3}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} for helping to determine the correct M\+SA file format.

Currently, the list of parsable multiple sequence alignment file formats consists of\+:
\begin{DoxyItemize}
\item \hyperlink{file_formats_msa-formats-clustal}{ClustalW format}
\item \hyperlink{file_formats_msa-formats-stockholm}{Stockholm 1.\+0 format}
\item \hyperlink{file_formats_msa-formats-fasta}{F\+A\+S\+TA (Pearson) format}
\item \hyperlink{file_formats_msa-formats-maf}{M\+AF format}
\end{DoxyItemize}\begin{DoxyNote}{Note}
After successfully reading an alignment, this function performs a validation of the data that includes uniqueness of the sequence identifiers, and equal sequence lengths. This check can be deactivated by passing \hyperlink{group__file__utils_ga229fb3778ecabea4782902b69fa48fd1}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK} in the {\ttfamily options} parameter.
\end{DoxyNote}
\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_gad02d5d12bda54611c915a1019323b7be}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, \hyperlink{group__file__utils_gade4fa8136ebb2d0f7eb3f8b59a8658e3}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()}, \hyperlink{group__file__utils_ga79a23de2c7249f2cccd762e475c81859}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL}, \hyperlink{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}, \hyperlink{group__file__utils_gacf6274a2c825f34e4131404665b00604}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA}, \hyperlink{group__file__utils_gab3186d774c01570d2a47d0308eec4927}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF}, \hyperlink{group__file__utils_ga4acc255373831856a8417b68de8a94c0}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}, \hyperlink{group__file__utils_ga229fb3778ecabea4782902b69fa48fd1}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK}
\end{DoxySeeAlso}

\begin{DoxyParams}{Parameters}
{\em fp} & The file pointer the data will be retrieved from \\
\hline
{\em names} & An address to the pointer where sequence identifiers should be written to \\
\hline
{\em aln} & An address to the pointer where aligned sequences should be written to \\
\hline
{\em id} & An address to the pointer where the alignment ID should be written to (Maybe N\+U\+LL) \\
\hline
{\em structure} & An address to the pointer where consensus structure information should be written to (Maybe N\+U\+LL) \\
\hline
{\em options} & Options to manipulate the behavior of this function \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
The number of sequences in the alignment, or -\/1 if no alignment record could be found 
\end{DoxyReturn}
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format@{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format}}
\index{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format@{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection[{\texorpdfstring{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format(const char $\ast$filename, unsigned int options)}{vrna_file_msa_detect_format(const char *filename, unsigned int options)}}]{\setlength{\rightskip}{0pt plus 5cm}unsigned int vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format (
\begin{DoxyParamCaption}
\item[{const char $\ast$}]{filename, }
\item[{unsigned int}]{options}
\end{DoxyParamCaption}
)}\hypertarget{group__file__utils_gade4fa8136ebb2d0f7eb3f8b59a8658e3}{}\label{group__file__utils_gade4fa8136ebb2d0f7eb3f8b59a8658e3}


{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Detect the format of a multiple sequence alignment file. 

This function attempts to determine the format of a file that supposedly contains a multiple sequence alignment (M\+SA). This is useful in cases where a M\+SA file contains more than a single record and therefore \hyperlink{group__file__utils_gad02d5d12bda54611c915a1019323b7be}{vrna\+\_\+file\+\_\+msa\+\_\+read()} can not be applied, since it only retrieves the first. Here, one can try to guess the correct file format using this function and then loop over the file, record by record using one of the low-\/level record retrieval functions for the corresponding M\+SA file format.

\begin{DoxyNote}{Note}
This function parses the entire first record within the specified file. As a result, it returns \hyperlink{group__file__utils_gabdc948f547e550125de3e7c65878400c}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN} not only if it can\textquotesingle{}t detect the file\textquotesingle{}s format, but also in cases where the file doesn\textquotesingle{}t contain sequences!
\end{DoxyNote}
\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_gad02d5d12bda54611c915a1019323b7be}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, vrna\+\_\+file\+\_\+stockholm\+\_\+read\+\_\+record(), vrna\+\_\+file\+\_\+clustal\+\_\+read\+\_\+record(), \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()}
\end{DoxySeeAlso}

\begin{DoxyParams}{Parameters}
{\em filename} & The name of input file that contains the alignment \\
\hline
{\em options} & Options to manipulate the behavior of this function \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
The M\+SA file format, or \hyperlink{group__file__utils_gabdc948f547e550125de3e7c65878400c}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN} 
\end{DoxyReturn}
