<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.11"/>
<title>RNAlib-2.2.10: Input / Output File Formats</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
  $(window).load(resizeHeight);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { init_search(); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">RNAlib-2.2.10
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.11 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li><a href="modules.html"><span>Modules</span></a></li>
      <li><a href="annotated.html"><span>Data&#160;Structures</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('file_formats.html','');});
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="headertitle">
<div class="title">Input / Output File Formats </div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><h1><a class="anchor" id="msa-formats"></a>
File formats for Multiple Sequence Alignments (MSA)</h1>
<h2><a class="anchor" id="msa-formats-clustal"></a>
ClustalW format</h2>
<p>The <em>ClustalW</em> format is a relatively simple text file containing a single multiple sequence alignment of DNA, RNA, or protein sequences. It was first used as an output format for the <em>clustalw</em> programs, but nowadays it may also be generated by various other sequence alignment tools. The specification is straight forward:</p>
<ul>
<li>The first line starts with the words<pre class="fragment">CLUSTAL W </pre> or <pre class="fragment">CLUSTALW </pre></li>
<li>After the above header there is at least one empty line</li>
<li>Finally, one or more blocks of sequence data are following, where each block is separated by at least one empty line</li>
</ul>
<p>Each line in a blocks of sequence data consists of the sequence name followed by the sequence symbols, separated by at least one whitespace character. Usually, the length of a sequence in one block does not exceed 60 symbols. Optionally, an additional whitespace separated cumulative residue count may follow the sequence symbols. Optionally, a block may be followed by a line depicting the degree of conservation of the respective alignment columns.</p>
<dl class="section note"><dt>Note</dt><dd>Sequence names and the sequences must not contain whitespace characters! Allowed gap symbols are the hyphen <em></em>("-"), and dot <em></em>(".").</dd></dl>
<dl class="section warning"><dt>Warning</dt><dd>Please note that many programs that output this format tend to truncate the sequence names to a limited number of characters, for instance the first 15 characters. This can destroy the uniqueness of identifiers in your MSA.</dd></dl>
<p>Here is an example alignment in ClustalW format: </p><pre class="fragment">CLUSTAL W (1.83) multiple sequence alignment


AL031296.1/85969-86120      CUGCCUCACAACGUUUGUGCCUCAGUUACCCGUAGAUGUAGUGAGGGUAACAAUACUUAC
AANU01225121.1/438-603      CUGCCUCACAACAUUUGUGCCUCAGUUACUCAUAGAUGUAGUGAGGGUGACAAUACUUAC
AAWR02037329.1/29294-29150  ---CUCGACACCACU---GCCUCGGUUACCCAUCGGUGCAGUGCGGGUAGUAGUACCAAU

AL031296.1/85969-86120      UCUCGUUGGUGAUAAGGAACAGCU
AANU01225121.1/438-603      UCUCGUUGGUGAUAAGGAACAGCU
AAWR02037329.1/29294-29150  GCUAAUUAGUUGUGAGGACCAACU </pre><h2><a class="anchor" id="msa-formats-stockholm"></a>
Stockholm 1.0 format</h2>
<p>Here is an example alignment in Stockholm 1.0 format: </p><pre class="fragment"># STOCKHOLM 1.0

#=GF AC   RF01293
#=GF ID   ACA59
#=GF DE   Small nucleolar RNA ACA59
#=GF AU   Wilkinson A
#=GF SE   Predicted; WAR; Wilkinson A
#=GF SS   Predicted; WAR; Wilkinson A
#=GF GA   43.00
#=GF TC   44.90
#=GF NC   40.30
#=GF TP   Gene; snRNA; snoRNA; HACA-box;
#=GF BM   cmbuild -F CM SEED
#=GF CB   cmcalibrate --mpi CM
#=GF SM   cmsearch --cpu 4 --verbose --nohmmonly -E 1000 -Z 549862.597050 CM SEQDB
#=GF DR   snoRNABase; ACA59;
#=GF DR   SO; 0001263; ncRNA_gene;
#=GF DR   GO; 0006396; RNA processing;
#=GF DR   GO; 0005730; nucleolus;
#=GF RN   [1]
#=GF RM   15199136
#=GF RT   Human box H/ACA pseudouridylation guide RNA machinery.
#=GF RA   Kiss AM, Jady BE, Bertrand E, Kiss T
#=GF RL   Mol Cell Biol. 2004;24:5797-5807.
#=GF WK   Small_nucleolar_RNA
#=GF SQ   3


AL031296.1/85969-86120     CUGCCUCACAACGUUUGUGCCUCAGUUACCCGUAGAUGUAGUGAGGGUAACAAUACUUACUCUCGUUGGUGAUAAGGAACAGCU
AANU01225121.1/438-603     CUGCCUCACAACAUUUGUGCCUCAGUUACUCAUAGAUGUAGUGAGGGUGACAAUACUUACUCUCGUUGGUGAUAAGGAACAGCU
AAWR02037329.1/29294-29150 ---CUCGACACCACU---GCCUCGGUUACCCAUCGGUGCAGUGCGGGUAGUAGUACCAAUGCUAAUUAGUUGUGAGGACCAACU
#=GC SS_cons               -----((((,&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;___________&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;,,,,&lt;&lt;&lt;&lt;&lt;&lt;&lt;______&gt;&gt;&gt;&gt;&gt;&gt;&gt;,,,,,))))::::::::::::
#=GC RF                    CUGCcccaCAaCacuuguGCCUCaGUUACcCauagguGuAGUGaGgGuggcAaUACccaCcCucgUUgGuggUaAGGAaCAgCU
//</pre><h2><a class="anchor" id="msa-formats-fasta"></a>
FASTA (Pearson) format</h2>
<dl class="section note"><dt>Note</dt><dd>Sequence names must not contain whitespace characters. Otherwise, the parts after the first whitespace will be dropped. The only allowed gap character is the hyphen <em></em>("-").</dd></dl>
<p>Here is an example alignment in FASTA format: </p><pre class="fragment">&gt;AL031296.1/85969-86120
CUGCCUCACAACGUUUGUGCCUCAGUUACCCGUAGAUGUAGUGAGGGUAACAAUACUUAC
UCUCGUUGGUGAUAAGGAACAGCU
&gt;AANU01225121.1/438-603
CUGCCUCACAACAUUUGUGCCUCAGUUACUCAUAGAUGUAGUGAGGGUGACAAUACUUAC
UCUCGUUGGUGAUAAGGAACAGCU
&gt;AAWR02037329.1/29294-29150
---CUCGACACCACU---GCCUCGGUUACCCAUCGGUGCAGUGCGGGUAGUAGUACCAAU
GCUAAUUAGUUGUGAGGACCAACU</pre><h2><a class="anchor" id="msa-formats-maf"></a>
MAF format</h2>
<p>The multiple alignment format (MAF) is usually used to store multiple alignments on DNA level between entire genomes. It consists of independent blocks of aligned sequences which are annotated by their genomic location. Consequently, an MAF formatted MSA file may contain multiple records. MAF files start with a line </p><pre class="fragment">##maf
</pre><p> which is optionally extended by whitespace delimited key=value pairs. Lines starting with the character ("#") are considered comments and usually ignored.</p>
<p>A MAF block starts with character ("a") at the beginning of a line, optionally followed by whitespace delimited key=value pairs. The next lines start with character ("s") and contain sequence information of the form </p><pre class="fragment">s src start size strand srcSize sequence
</pre><p> where</p><ul>
<li><em>src</em> is the name of the sequence source</li>
<li><em>start</em> is the start of the aligned region within the source (0-based)</li>
<li><em>size</em> is the length of the aligned region without gap characters</li>
<li><em>strand</em> is either ("+") or ("-"), depicting the location of the aligned region relative to the source</li>
<li><em>srcSize</em> is the size of the entire sequence source, e.g. the full chromosome</li>
<li><em>sequence</em> is the aligned sequence including gaps depicted by the hyphen ("-")</li>
</ul>
<p>Here is an example alignment in MAF format (bluntly taken from the <a href="https://cgwb.nci.nih.gov/FAQ/FAQformat.html#format5">UCSC Genome browser website</a>): </p><pre class="fragment">##maf version=1 scoring=tba.v8 
# tba.v8 (((human chimp) baboon) (mouse rat)) 
# multiz.v7
# maf_project.v5 _tba_right.maf3 mouse _tba_C
# single_cov2.v4 single_cov2 /dev/stdin
                   
a score=23262.0     
s hg16.chr7    27578828 38 + 158545518 AAA-GGGAATGTTAACCAAATGA---ATTGTCTCTTACGGTG
s panTro1.chr6 28741140 38 + 161576975 AAA-GGGAATGTTAACCAAATGA---ATTGTCTCTTACGGTG
s baboon         116834 38 +   4622798 AAA-GGGAATGTTAACCAAATGA---GTTGTCTCTTATGGTG
s mm4.chr6     53215344 38 + 151104725 -AATGGGAATGTTAAGCAAACGA---ATTGTCTCTCAGTGTG
s rn3.chr4     81344243 40 + 187371129 -AA-GGGGATGCTAAGCCAATGAGTTGTTGTCTCTCAATGTG
                   
a score=5062.0                    
s hg16.chr7    27699739 6 + 158545518 TAAAGA
s panTro1.chr6 28862317 6 + 161576975 TAAAGA
s baboon         241163 6 +   4622798 TAAAGA 
s mm4.chr6     53303881 6 + 151104725 TAAAGA
s rn3.chr4     81444246 6 + 187371129 taagga

a score=6636.0
s hg16.chr7    27707221 13 + 158545518 gcagctgaaaaca
s panTro1.chr6 28869787 13 + 161576975 gcagctgaaaaca
s baboon         249182 13 +   4622798 gcagctgaaaaca
s mm4.chr6     53310102 13 + 151104725 ACAGCTGAAAATA</pre><h1><a class="anchor" id="constraint-formats"></a>
File formats for Secondary Structure Constraints</h1>
<h2><a class="anchor" id="constraint-formats-file"></a>
Constraints Definition File</h2>
<p>The RNAlib can parse and apply data from constraint definition text files, where each constraint is given as a line of whitespace delimited commands. The syntax we use extends the one used in <a href="http://mfold.rna.albany.edu/?q=mfold">mfold</a> / <a href="http://mfold.rna.albany.edu/?q=DINAMelt/software">UNAfold</a> where each line begins with a command character followed by a set of positions.<br />
 Additionally, we introduce several new commands, and allow for an optional loop type context specifier in form of a sequence of characters, and an orientation flag that enables one to force a nucleotide to pair upstream, or downstream.</p>
<h3><a class="anchor" id="const_file_commands"></a>
Constraint commands</h3>
<p>The following set of commands is recognized:</p><ul>
<li><code>F</code> <img class="formulaInl" alt="$ \ldots $" src="form_1.png"/> Force</li>
<li><code>P</code> <img class="formulaInl" alt="$ \ldots $" src="form_1.png"/> Prohibit</li>
<li><code>C</code> <img class="formulaInl" alt="$ \ldots $" src="form_1.png"/> Conflicts/Context dependency</li>
<li><code>A</code> <img class="formulaInl" alt="$ \ldots $" src="form_1.png"/> Allow (for non-canonical pairs)</li>
<li><code>E</code> <img class="formulaInl" alt="$ \ldots $" src="form_1.png"/> Soft constraints for unpaired position(s), or base pair(s)</li>
</ul>
<h3><a class="anchor" id="const_file_loop_types"></a>
Specification of the loop type context</h3>
<p>The optional loop type context specifier <code></code>[WHERE] may be a combination of the following:</p><ul>
<li><code>E</code> <img class="formulaInl" alt="$ \ldots $" src="form_1.png"/> Exterior loop</li>
<li><code>H</code> <img class="formulaInl" alt="$ \ldots $" src="form_1.png"/> Hairpin loop</li>
<li><code>I</code> <img class="formulaInl" alt="$ \ldots $" src="form_1.png"/> Interior loop (enclosing pair)</li>
<li><code>i</code> <img class="formulaInl" alt="$ \ldots $" src="form_1.png"/> Interior loop (enclosed pair)</li>
<li><code>M</code> <img class="formulaInl" alt="$ \ldots $" src="form_1.png"/> Multibranch loop (enclosing pair)</li>
<li><code>m</code> <img class="formulaInl" alt="$ \ldots $" src="form_1.png"/> Multibranch loop (enclosed pair)</li>
<li><code>A</code> <img class="formulaInl" alt="$ \ldots $" src="form_1.png"/> All loops</li>
</ul>
<p>If no <code></code>[WHERE] flags are set, all contexts are considered (equivalent to <code>A</code> )</p>
<h3><a class="anchor" id="const_file_orientation"></a>
Controlling the orientation of base pairing</h3>
<p>For particular nucleotides that are forced to pair, the following <code></code>[ORIENTATION] flags may be used:</p><ul>
<li><code>U</code> <img class="formulaInl" alt="$ \ldots $" src="form_1.png"/> Upstream</li>
<li><code>D</code> <img class="formulaInl" alt="$ \ldots $" src="form_1.png"/> Downstream</li>
</ul>
<p>If no <code></code>[ORIENTATION] flag is set, both directions are considered.</p>
<h3><a class="anchor" id="const_file_seq_coords"></a>
Sequence coordinates</h3>
<p>Sequence positions of nucleotides/base pairs are <img class="formulaInl" alt="$ 1- $" src="form_2.png"/> based and consist of three positions <img class="formulaInl" alt="$ i $" src="form_3.png"/>, <img class="formulaInl" alt="$ j $" src="form_4.png"/>, and <img class="formulaInl" alt="$ k $" src="form_5.png"/>. Alternativly, four positions may be provided as a pair of two position ranges <img class="formulaInl" alt="$ [i:j] $" src="form_6.png"/>, and <img class="formulaInl" alt="$ [k:l] $" src="form_7.png"/> using the '-' sign as delimiter within each range, i.e. <img class="formulaInl" alt="$ i-j $" src="form_8.png"/>, and <img class="formulaInl" alt="$ k-l $" src="form_9.png"/>.</p>
<h3><a class="anchor" id="const_file_syntax"></a>
Valid constraint commands</h3>
<p>Below are resulting general cases that are considered <em>valid</em> constraints:</p>
<ol type="1">
<li><b>"Forcing a range of nucleotide positions to be paired"</b>:<br />
 Syntax:<div class="fragment"><div class="line">F i 0 k [WHERE] [ORIENTATION] </div></div><!-- fragment --><br />
 Description:<br />
 Enforces the set of <img class="formulaInl" alt="$ k $" src="form_5.png"/> consecutive nucleotides starting at position <img class="formulaInl" alt="$ i $" src="form_3.png"/> to be paired. The optional loop type specifier <code></code>[WHERE] allows to force them to appear as closing/enclosed pairs of certain types of loops.</li>
<li><b>"Forcing a set of consecutive base pairs to form"</b>:<br />
 Syntax:<pre class="fragment">F i j k [WHERE] </pre><br />
 Description:<br />
 Enforces the base pairs <img class="formulaInl" alt="$ (i,j), \ldots, (i+(k-1), j-(k-1)) $" src="form_10.png"/> to form. The optional loop type specifier <code></code>[WHERE] allows to specify in which loop context the base pair must appear.</li>
<li><b>"Prohibiting a range of nucleotide positions to be paired"</b>:<br />
 Syntax:<pre class="fragment">P i 0 k [WHERE] </pre><br />
 Description:<br />
 Prohibit a set of <img class="formulaInl" alt="$ k $" src="form_5.png"/> consecutive nucleotides to participate in base pairing, i.e. make these positions unpaired. The optional loop type specifier <code></code>[WHERE] allows to force the nucleotides to appear within the loop of specific types.</li>
<li><b>"Probibiting a set of consecutive base pairs to form"</b>:<br />
 Syntax:<pre class="fragment">P i j k [WHERE] </pre><br />
 Description:<br />
 Probibit the base pairs <img class="formulaInl" alt="$ (i,j), \ldots, (i+(k-1), j-(k-1)) $" src="form_10.png"/> to form. The optional loop type specifier <code></code>[WHERE] allows to specify the type of loop they are disallowed to be the closing or an enclosed pair of.</li>
<li><b>"Prohibiting two ranges of nucleotides to pair with each other"</b>:<br />
 Syntax:<pre class="fragment">P i-j k-l [WHERE] </pre> Description:<br />
 Prohibit any nucleotide <img class="formulaInl" alt="$ p \in [i:j] $" src="form_11.png"/> to pair with any other nucleotide <img class="formulaInl" alt="$ q \in [k:l] $" src="form_12.png"/>. The optional loop type specifier <code></code>[WHERE] allows to specify the type of loop they are disallowed to be the closing or an enclosed pair of.</li>
<li><b>"Enforce a loop context for a range of nucleotide positions"</b>:<br />
 Syntax:<pre class="fragment">C i 0 k [WHERE] </pre> Description:<br />
 This command enforces nucleotides to be unpaired similar to <em>prohibiting</em> nucleotides to be paired, as described above. It too marks the corresponding nucleotides to be unpaired, however, the <code></code>[WHERE] flag can be used to enforce specfic loop types the nucleotides must appear in.</li>
<li><b>"Remove pairs that conflict with a set of consecutive base pairs"</b>:<br />
 Syntax:<pre class="fragment">C i j k </pre><br />
 Description:<br />
 Remove all base pairs that conflict with a set of consecutive base pairs <img class="formulaInl" alt="$ (i,j), \ldots, (i+(k-1), j-(k-1)) $" src="form_10.png"/>. Two base pairs <img class="formulaInl" alt="$ (i,j) $" src="form_13.png"/> and <img class="formulaInl" alt="$ (p,q) $" src="form_14.png"/> conflict with each other if <img class="formulaInl" alt="$ i &lt; p &lt; j &lt; q $" src="form_15.png"/>, or <img class="formulaInl" alt="$ p &lt; i &lt; q &lt; j $" src="form_16.png"/>.</li>
<li><b>"Allow a set of consecutive (non-canonical) base pairs to form"</b>:<br />
 Syntax:<div class="fragment"><div class="line">A i j k [WHERE] </div></div><!-- fragment --><br />
 Description:<br />
 This command enables the formation of the consecutive base pairs <img class="formulaInl" alt="$ (i,j), \ldots, (i+(k-1), j-(k-1)) $" src="form_10.png"/>, no matter if they are <em>canonical</em>, or <em>non-canonical</em>. In contrast to the above <code>F</code> and <code>W</code> commands, which remove conflicting base pairs, the <code>A</code> command does not. Therefore, it may be used to allow <em>non-canoncial</em> base pair interactions. Since the RNAlib does not contain free energy contributions <img class="formulaInl" alt="$ E_{ij} $" src="form_17.png"/> for non-canonical base pairs <img class="formulaInl" alt="$ (i,j) $" src="form_13.png"/>, they are scored as the <em>maximum</em> of similar, known contributions. In terms of a <em>Nussinov</em> like scoring function the free energy of non-canonical base pairs is therefore estimated as <p class="formulaDsp">
<img class="formulaDsp" alt="\[ E_{ij} = \min \left[ \max_{(i,k) \in \{GC, CG, AU, UA, GU, UG\}} E_{ik}, \max_{(k,j) \in \{GC, CG, AU, UA, GU, UG\}} E_{kj} \right]. \]" src="form_18.png"/>
</p>
 The optional loop type specifier <code></code>[WHERE] allows to specify in which loop context the base pair may appear.</li>
<li><b>"Apply pseudo free energy to a range of unpaired nucleotide positions"</b>:<br />
 Syntax:<div class="fragment"><div class="line">E i 0 k e </div></div><!-- fragment --><br />
 Description:<br />
 Use this command to apply a pseudo free energy of <img class="formulaInl" alt="$ e $" src="form_19.png"/> to the set of <img class="formulaInl" alt="$ k $" src="form_5.png"/> consecutive nucleotides, starting at position <img class="formulaInl" alt="$ i $" src="form_3.png"/>. The pseudo free energy is applied only if these nucleotides are considered unpaired in the recursions, or evaluations, and is expected to be given in <img class="formulaInl" alt="$ kcal / mol $" src="form_20.png"/>.</li>
<li><b>"Apply pseudo free energy to a set of consecutive base pairs"</b>:<br />
 Syntax<div class="fragment"><div class="line">E i j k e </div></div><!-- fragment --><br />
 Use this command to apply a pseudo free energy of <img class="formulaInl" alt="$ e $" src="form_19.png"/> to the set of base pairs <img class="formulaInl" alt="$ (i,j), \ldots, (i+(k-1), j-(k-1)) $" src="form_10.png"/>. Energies are expected to be given in <img class="formulaInl" alt="$ kcal / mol $" src="form_20.png"/>. </li>
</ol>
</div></div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated on Tue Sep 6 2016 15:07:07 for RNAlib-2.2.10 by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.11 </li>
  </ul>
</div>
</body>
</html>
