<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.11"/>
<title>RNAlib-2.2.10: Functions to Read/Write several File Formats for RNA Sequences, Structures, and Alignments</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
  $(window).load(resizeHeight);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { init_search(); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">RNAlib-2.2.10
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.11 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li class="current"><a href="modules.html"><span>Modules</span></a></li>
      <li><a href="annotated.html"><span>Data&#160;Structures</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('group__file__utils.html','');});
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#files">Files</a> &#124;
<a href="#define-members">Macros</a> &#124;
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle">
<div class="title">Functions to Read/Write several File Formats for RNA Sequences, Structures, and Alignments<div class="ingroups"><a class="el" href="group__utils.html">Utilities</a></div></div>  </div>
</div><!--header-->
<div class="contents">
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Collaboration diagram for Functions to Read/Write several File Formats for RNA Sequences, Structures, and Alignments:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<center><table><tr><td><div class="center"><iframe scrolling="no" frameborder="0" src="group__file__utils.svg" width="340" height="82"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe>
</div>
</td></tr></table></center>
</div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="files"></a>
Files</h2></td></tr>
<tr class="memitem:file__formats_8h"><td class="memItemLeft" align="right" valign="top">file &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="file__formats_8h.html">file_formats.h</a></td></tr>
<tr class="memdesc:file__formats_8h"><td class="mdescLeft">&#160;</td><td class="mdescRight">Functions dealing with file formats for RNA sequences, and structures. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:file__formats__msa_8h"><td class="memItemLeft" align="right" valign="top">file &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="file__formats__msa_8h.html">file_formats_msa.h</a></td></tr>
<tr class="memdesc:file__formats__msa_8h"><td class="mdescLeft">&#160;</td><td class="mdescRight">Functions dealing with file formats for Multiple Sequence Alignments (MSA) <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ribo_8h"><td class="memItemLeft" align="right" valign="top">file &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="ribo_8h.html">ribo.h</a></td></tr>
<tr class="memdesc:ribo_8h"><td class="mdescLeft">&#160;</td><td class="mdescRight">Parse RiboSum Scoring Matrices for Covariance Scoring of Alignments. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="define-members"></a>
Macros</h2></td></tr>
<tr class="memitem:gabec89c09874528c6cb73140a4c3d86d7"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#gabec89c09874528c6cb73140a4c3d86d7">VRNA_OPTION_MULTILINE</a>&#160;&#160;&#160;32U</td></tr>
<tr class="memdesc:gabec89c09874528c6cb73140a4c3d86d7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Tell a function that an input is assumed to span several lines.  <a href="#gabec89c09874528c6cb73140a4c3d86d7">More...</a><br /></td></tr>
<tr class="separator:gabec89c09874528c6cb73140a4c3d86d7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7d725ef525b29891abef3f1ed42599a4"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#ga7d725ef525b29891abef3f1ed42599a4">VRNA_CONSTRAINT_MULTILINE</a>&#160;&#160;&#160;32U</td></tr>
<tr class="memdesc:ga7d725ef525b29891abef3f1ed42599a4"><td class="mdescLeft">&#160;</td><td class="mdescRight">parse multiline constraint  <a href="#ga7d725ef525b29891abef3f1ed42599a4">More...</a><br /></td></tr>
<tr class="separator:ga7d725ef525b29891abef3f1ed42599a4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga79a23de2c7249f2cccd762e475c81859"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#ga79a23de2c7249f2cccd762e475c81859">VRNA_FILE_FORMAT_MSA_CLUSTAL</a>&#160;&#160;&#160;1U</td></tr>
<tr class="memdesc:ga79a23de2c7249f2cccd762e475c81859"><td class="mdescLeft">&#160;</td><td class="mdescRight">Option flag indicating ClustalW formatted files.  <a href="#ga79a23de2c7249f2cccd762e475c81859">More...</a><br /></td></tr>
<tr class="separator:ga79a23de2c7249f2cccd762e475c81859"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga62be992445cd8ab2ad7a8fded944338b"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#ga62be992445cd8ab2ad7a8fded944338b">VRNA_FILE_FORMAT_MSA_STOCKHOLM</a>&#160;&#160;&#160;2U</td></tr>
<tr class="memdesc:ga62be992445cd8ab2ad7a8fded944338b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Option flag indicating Stockholm 1.0 formatted files.  <a href="#ga62be992445cd8ab2ad7a8fded944338b">More...</a><br /></td></tr>
<tr class="separator:ga62be992445cd8ab2ad7a8fded944338b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gacf6274a2c825f34e4131404665b00604"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#gacf6274a2c825f34e4131404665b00604">VRNA_FILE_FORMAT_MSA_FASTA</a>&#160;&#160;&#160;4U</td></tr>
<tr class="memdesc:gacf6274a2c825f34e4131404665b00604"><td class="mdescLeft">&#160;</td><td class="mdescRight">Option flag indicating FASTA (Pearson) formatted files.  <a href="#gacf6274a2c825f34e4131404665b00604">More...</a><br /></td></tr>
<tr class="separator:gacf6274a2c825f34e4131404665b00604"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab3186d774c01570d2a47d0308eec4927"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#gab3186d774c01570d2a47d0308eec4927">VRNA_FILE_FORMAT_MSA_MAF</a>&#160;&#160;&#160;8U</td></tr>
<tr class="memdesc:gab3186d774c01570d2a47d0308eec4927"><td class="mdescLeft">&#160;</td><td class="mdescRight">Option flag indicating MAF formatted files.  <a href="#gab3186d774c01570d2a47d0308eec4927">More...</a><br /></td></tr>
<tr class="separator:gab3186d774c01570d2a47d0308eec4927"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga4acc255373831856a8417b68de8a94c0"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#ga4acc255373831856a8417b68de8a94c0">VRNA_FILE_FORMAT_MSA_DEFAULT</a></td></tr>
<tr class="memdesc:ga4acc255373831856a8417b68de8a94c0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Option flag indicating the set of default file formats.  <a href="#ga4acc255373831856a8417b68de8a94c0">More...</a><br /></td></tr>
<tr class="separator:ga4acc255373831856a8417b68de8a94c0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga229fb3778ecabea4782902b69fa48fd1"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#ga229fb3778ecabea4782902b69fa48fd1">VRNA_FILE_FORMAT_MSA_NOCHECK</a>&#160;&#160;&#160;4096U</td></tr>
<tr class="memdesc:ga229fb3778ecabea4782902b69fa48fd1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Option flag to disable validation of the alignment.  <a href="#ga229fb3778ecabea4782902b69fa48fd1">More...</a><br /></td></tr>
<tr class="separator:ga229fb3778ecabea4782902b69fa48fd1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabdc948f547e550125de3e7c65878400c"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#gabdc948f547e550125de3e7c65878400c">VRNA_FILE_FORMAT_MSA_UNKNOWN</a>&#160;&#160;&#160;8192U</td></tr>
<tr class="memdesc:gabdc948f547e550125de3e7c65878400c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return flag of <a class="el" href="group__file__utils.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file. ">vrna_file_msa_detect_format()</a> to indicate unknown or malformatted alignment.  <a href="#gabdc948f547e550125de3e7c65878400c">More...</a><br /></td></tr>
<tr class="separator:gabdc948f547e550125de3e7c65878400c"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:gaaface7db12fadc3d271641c4515ab6e4"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#gaaface7db12fadc3d271641c4515ab6e4">vrna_file_helixlist</a> (const char *seq, const char *db, float energy, FILE *file)</td></tr>
<tr class="memdesc:gaaface7db12fadc3d271641c4515ab6e4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print a secondary structure as helix list.  <a href="#gaaface7db12fadc3d271641c4515ab6e4">More...</a><br /></td></tr>
<tr class="separator:gaaface7db12fadc3d271641c4515ab6e4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab69682373ccca1e0e28cc967eec07745"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#gab69682373ccca1e0e28cc967eec07745">vrna_file_connect</a> (const char *seq, const char *db, float energy, const char *identifier, FILE *file)</td></tr>
<tr class="memdesc:gab69682373ccca1e0e28cc967eec07745"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print a secondary structure as connect table.  <a href="#gab69682373ccca1e0e28cc967eec07745">More...</a><br /></td></tr>
<tr class="separator:gab69682373ccca1e0e28cc967eec07745"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9b462e6f202594af5d3fa56e280d633f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#ga9b462e6f202594af5d3fa56e280d633f">vrna_file_bpseq</a> (const char *seq, const char *db, FILE *file)</td></tr>
<tr class="memdesc:ga9b462e6f202594af5d3fa56e280d633f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print a secondary structure in bpseq format.  <a href="#ga9b462e6f202594af5d3fa56e280d633f">More...</a><br /></td></tr>
<tr class="separator:ga9b462e6f202594af5d3fa56e280d633f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga31f4a6c2ea1495a6e4f9eb45a9f6193d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#ga31f4a6c2ea1495a6e4f9eb45a9f6193d">vrna_file_json</a> (const char *seq, const char *db, double energy, const char *identifier, FILE *file)</td></tr>
<tr class="memdesc:ga31f4a6c2ea1495a6e4f9eb45a9f6193d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print a secondary structure in jsonformat.  <a href="#ga31f4a6c2ea1495a6e4f9eb45a9f6193d">More...</a><br /></td></tr>
<tr class="separator:ga31f4a6c2ea1495a6e4f9eb45a9f6193d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8cfb7e271efc9e1f34640acb85475639"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#ga8cfb7e271efc9e1f34640acb85475639">vrna_file_fasta_read_record</a> (char **header, char **sequence, char ***rest, FILE *file, unsigned int options)</td></tr>
<tr class="memdesc:ga8cfb7e271efc9e1f34640acb85475639"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get a (fasta) data set from a file or stdin.  <a href="#ga8cfb7e271efc9e1f34640acb85475639">More...</a><br /></td></tr>
<tr class="separator:ga8cfb7e271efc9e1f34640acb85475639"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad37cbb63a05eed63ba25c91628409be0"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#gad37cbb63a05eed63ba25c91628409be0">vrna_extract_record_rest_structure</a> (const char **lines, unsigned int length, unsigned int option)</td></tr>
<tr class="memdesc:gad37cbb63a05eed63ba25c91628409be0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Extract a dot-bracket structure string from (multiline)character array.  <a href="#gad37cbb63a05eed63ba25c91628409be0">More...</a><br /></td></tr>
<tr class="separator:gad37cbb63a05eed63ba25c91628409be0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga646ebf45450a69a7f2533f9ecd283a32"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#ga646ebf45450a69a7f2533f9ecd283a32">vrna_file_SHAPE_read</a> (const char *file_name, int length, double default_value, char *sequence, double *values)</td></tr>
<tr class="memdesc:ga646ebf45450a69a7f2533f9ecd283a32"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read data from a given SHAPE reactivity input file.  <a href="#ga646ebf45450a69a7f2533f9ecd283a32">More...</a><br /></td></tr>
<tr class="separator:ga646ebf45450a69a7f2533f9ecd283a32"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae33323c53765ecbbc410d9de2d495432"><td class="memItemLeft" align="right" valign="top"><a class="el" href="group__data__structures.html#ga8e4eb5e1bfc95776559575beb359af87">vrna_plist_t</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#gae33323c53765ecbbc410d9de2d495432">vrna_file_constraints_read</a> (const char *filename, unsigned int length, unsigned int options)</td></tr>
<tr class="memdesc:gae33323c53765ecbbc410d9de2d495432"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read constraints from an input file.  <a href="#gae33323c53765ecbbc410d9de2d495432">More...</a><br /></td></tr>
<tr class="separator:gae33323c53765ecbbc410d9de2d495432"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga55a9ae6dfeecc1b3f0c2acf6fa796c15"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#ga55a9ae6dfeecc1b3f0c2acf6fa796c15">vrna_extract_record_rest_constraint</a> (char **cstruc, const char **lines, unsigned int option)</td></tr>
<tr class="memdesc:ga55a9ae6dfeecc1b3f0c2acf6fa796c15"><td class="mdescLeft">&#160;</td><td class="mdescRight">Extract a hard constraint encoded as pseudo dot-bracket string.  <a href="#ga55a9ae6dfeecc1b3f0c2acf6fa796c15">More...</a><br /></td></tr>
<tr class="separator:ga55a9ae6dfeecc1b3f0c2acf6fa796c15"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gafd194a69af9d92b5b0412a7627ac1595"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#gafd194a69af9d92b5b0412a7627ac1595">read_record</a> (char **header, char **sequence, char ***rest, unsigned int options)</td></tr>
<tr class="memdesc:gafd194a69af9d92b5b0412a7627ac1595"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get a data record from stdin.  <a href="#gafd194a69af9d92b5b0412a7627ac1595">More...</a><br /></td></tr>
<tr class="separator:gafd194a69af9d92b5b0412a7627ac1595"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad02d5d12bda54611c915a1019323b7be"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#gad02d5d12bda54611c915a1019323b7be">vrna_file_msa_read</a> (const char *filename, char ***names, char ***aln, char **id, char **structure, unsigned int options)</td></tr>
<tr class="memdesc:gad02d5d12bda54611c915a1019323b7be"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read a multiple sequence alignment from file.  <a href="#gad02d5d12bda54611c915a1019323b7be">More...</a><br /></td></tr>
<tr class="separator:gad02d5d12bda54611c915a1019323b7be"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad4203a438622b2df7bc2f16578d54799"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#gad4203a438622b2df7bc2f16578d54799">vrna_file_msa_read_record</a> (FILE *fp, char ***names, char ***aln, char **id, char **structure, unsigned int options)</td></tr>
<tr class="memdesc:gad4203a438622b2df7bc2f16578d54799"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read a multiple sequence alignment from file handle.  <a href="#gad4203a438622b2df7bc2f16578d54799">More...</a><br /></td></tr>
<tr class="separator:gad4203a438622b2df7bc2f16578d54799"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gade4fa8136ebb2d0f7eb3f8b59a8658e3"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3">vrna_file_msa_detect_format</a> (const char *filename, unsigned int options)</td></tr>
<tr class="memdesc:gade4fa8136ebb2d0f7eb3f8b59a8658e3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Detect the format of a multiple sequence alignment file.  <a href="#gade4fa8136ebb2d0f7eb3f8b59a8658e3">More...</a><br /></td></tr>
<tr class="separator:gade4fa8136ebb2d0f7eb3f8b59a8658e3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5e125c9586fcd4e2e1559fe76f7289cc"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ga5e125c9586fcd4e2e1559fe76f7289cc"></a>
float **&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__utils.html#ga5e125c9586fcd4e2e1559fe76f7289cc">readribosum</a> (char *name)</td></tr>
<tr class="memdesc:ga5e125c9586fcd4e2e1559fe76f7289cc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read a RiboSum or other user-defined Scoring Matrix and Store into global Memory. <br /></td></tr>
<tr class="separator:ga5e125c9586fcd4e2e1559fe76f7289cc"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<h2 class="groupheader">Macro Definition Documentation</h2>
<a class="anchor" id="gabec89c09874528c6cb73140a4c3d86d7"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_OPTION_MULTILINE&#160;&#160;&#160;32U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats_8h.html">ViennaRNA/file_formats.h</a>&gt;</code></p>

<p>Tell a function that an input is assumed to span several lines. </p>
<p>If used as input-option a function might also be returning this state telling that it has read data from multiple lines.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__utils.html#gad37cbb63a05eed63ba25c91628409be0" title="Extract a dot-bracket structure string from (multiline)character array. ">vrna_extract_record_rest_structure()</a>, <a class="el" href="group__file__utils.html#ga8cfb7e271efc9e1f34640acb85475639" title="Get a (fasta) data set from a file or stdin. ">vrna_file_fasta_read_record()</a> </dd></dl>

</div>
</div>
<a class="anchor" id="ga7d725ef525b29891abef3f1ed42599a4"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_CONSTRAINT_MULTILINE&#160;&#160;&#160;32U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats_8h.html">ViennaRNA/file_formats.h</a>&gt;</code></p>

<p>parse multiline constraint </p>
<dl class="deprecated"><dt><b><a class="el" href="deprecated.html#_deprecated000065">Deprecated:</a></b></dt><dd>see <a class="el" href="group__file__utils.html#gad37cbb63a05eed63ba25c91628409be0" title="Extract a dot-bracket structure string from (multiline)character array. ">vrna_extract_record_rest_structure()</a> </dd></dl>

</div>
</div>
<a class="anchor" id="ga79a23de2c7249f2cccd762e475c81859"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_CLUSTAL&#160;&#160;&#160;1U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats__msa_8h.html">ViennaRNA/file_formats_msa.h</a>&gt;</code></p>

<p>Option flag indicating ClustalW formatted files. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__utils.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file. ">vrna_file_msa_read()</a>, <a class="el" href="group__file__utils.html#gad4203a438622b2df7bc2f16578d54799" title="Read a multiple sequence alignment from file handle. ">vrna_file_msa_read_record()</a>, <a class="el" href="group__file__utils.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file. ">vrna_file_msa_detect_format()</a> </dd></dl>

</div>
</div>
<a class="anchor" id="ga62be992445cd8ab2ad7a8fded944338b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_STOCKHOLM&#160;&#160;&#160;2U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats__msa_8h.html">ViennaRNA/file_formats_msa.h</a>&gt;</code></p>

<p>Option flag indicating Stockholm 1.0 formatted files. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__utils.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file. ">vrna_file_msa_read()</a>, <a class="el" href="group__file__utils.html#gad4203a438622b2df7bc2f16578d54799" title="Read a multiple sequence alignment from file handle. ">vrna_file_msa_read_record()</a>, <a class="el" href="group__file__utils.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file. ">vrna_file_msa_detect_format()</a> </dd></dl>

</div>
</div>
<a class="anchor" id="gacf6274a2c825f34e4131404665b00604"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_FASTA&#160;&#160;&#160;4U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats__msa_8h.html">ViennaRNA/file_formats_msa.h</a>&gt;</code></p>

<p>Option flag indicating FASTA (Pearson) formatted files. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__utils.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file. ">vrna_file_msa_read()</a>, <a class="el" href="group__file__utils.html#gad4203a438622b2df7bc2f16578d54799" title="Read a multiple sequence alignment from file handle. ">vrna_file_msa_read_record()</a>, <a class="el" href="group__file__utils.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file. ">vrna_file_msa_detect_format()</a> </dd></dl>

</div>
</div>
<a class="anchor" id="gab3186d774c01570d2a47d0308eec4927"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_MAF&#160;&#160;&#160;8U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats__msa_8h.html">ViennaRNA/file_formats_msa.h</a>&gt;</code></p>

<p>Option flag indicating MAF formatted files. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__utils.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file. ">vrna_file_msa_read()</a>, <a class="el" href="group__file__utils.html#gad4203a438622b2df7bc2f16578d54799" title="Read a multiple sequence alignment from file handle. ">vrna_file_msa_read_record()</a>, <a class="el" href="group__file__utils.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file. ">vrna_file_msa_detect_format()</a> </dd></dl>

</div>
</div>
<a class="anchor" id="ga4acc255373831856a8417b68de8a94c0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_DEFAULT</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats__msa_8h.html">ViennaRNA/file_formats_msa.h</a>&gt;</code></p>
<b>Value:</b><div class="fragment"><div class="line">( \</div><div class="line">                                              VRNA_FILE_FORMAT_MSA_CLUSTAL \</div><div class="line">                                            | <a class="code" href="group__file__utils.html#ga62be992445cd8ab2ad7a8fded944338b">VRNA_FILE_FORMAT_MSA_STOCKHOLM</a> \</div><div class="line">                                            | <a class="code" href="group__file__utils.html#gacf6274a2c825f34e4131404665b00604">VRNA_FILE_FORMAT_MSA_FASTA</a> \</div><div class="line">                                            | <a class="code" href="group__file__utils.html#gab3186d774c01570d2a47d0308eec4927">VRNA_FILE_FORMAT_MSA_MAF</a> \</div><div class="line">                                          )</div><div class="ttc" id="group__file__utils_html_gab3186d774c01570d2a47d0308eec4927"><div class="ttname"><a href="group__file__utils.html#gab3186d774c01570d2a47d0308eec4927">VRNA_FILE_FORMAT_MSA_MAF</a></div><div class="ttdeci">#define VRNA_FILE_FORMAT_MSA_MAF</div><div class="ttdoc">Option flag indicating MAF formatted files. </div><div class="ttdef"><b>Definition:</b> file_formats_msa.h:38</div></div>
<div class="ttc" id="group__file__utils_html_gacf6274a2c825f34e4131404665b00604"><div class="ttname"><a href="group__file__utils.html#gacf6274a2c825f34e4131404665b00604">VRNA_FILE_FORMAT_MSA_FASTA</a></div><div class="ttdeci">#define VRNA_FILE_FORMAT_MSA_FASTA</div><div class="ttdoc">Option flag indicating FASTA (Pearson) formatted files. </div><div class="ttdef"><b>Definition:</b> file_formats_msa.h:32</div></div>
<div class="ttc" id="group__file__utils_html_ga62be992445cd8ab2ad7a8fded944338b"><div class="ttname"><a href="group__file__utils.html#ga62be992445cd8ab2ad7a8fded944338b">VRNA_FILE_FORMAT_MSA_STOCKHOLM</a></div><div class="ttdeci">#define VRNA_FILE_FORMAT_MSA_STOCKHOLM</div><div class="ttdoc">Option flag indicating Stockholm 1.0 formatted files. </div><div class="ttdef"><b>Definition:</b> file_formats_msa.h:26</div></div>
</div><!-- fragment -->
<p>Option flag indicating the set of default file formats. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__utils.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file. ">vrna_file_msa_read()</a>, <a class="el" href="group__file__utils.html#gad4203a438622b2df7bc2f16578d54799" title="Read a multiple sequence alignment from file handle. ">vrna_file_msa_read_record()</a>, <a class="el" href="group__file__utils.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file. ">vrna_file_msa_detect_format()</a> </dd></dl>

</div>
</div>
<a class="anchor" id="ga229fb3778ecabea4782902b69fa48fd1"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_NOCHECK&#160;&#160;&#160;4096U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats__msa_8h.html">ViennaRNA/file_formats_msa.h</a>&gt;</code></p>

<p>Option flag to disable validation of the alignment. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__utils.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file. ">vrna_file_msa_read()</a>, <a class="el" href="group__file__utils.html#gad4203a438622b2df7bc2f16578d54799" title="Read a multiple sequence alignment from file handle. ">vrna_file_msa_read_record()</a> </dd></dl>

</div>
</div>
<a class="anchor" id="gabdc948f547e550125de3e7c65878400c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_UNKNOWN&#160;&#160;&#160;8192U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats__msa_8h.html">ViennaRNA/file_formats_msa.h</a>&gt;</code></p>

<p>Return flag of <a class="el" href="group__file__utils.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file. ">vrna_file_msa_detect_format()</a> to indicate unknown or malformatted alignment. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__utils.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file. ">vrna_file_msa_detect_format()</a> </dd></dl>

</div>
</div>
<h2 class="groupheader">Function Documentation</h2>
<a class="anchor" id="gaaface7db12fadc3d271641c4515ab6e4"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vrna_file_helixlist </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>seq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>db</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>energy</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats_8h.html">ViennaRNA/file_formats.h</a>&gt;</code></p>

<p>Print a secondary structure as helix list. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">seq</td><td>The RNA sequence </td></tr>
    <tr><td class="paramname">db</td><td>The structure in dot-bracket format </td></tr>
    <tr><td class="paramname">energy</td><td>Free energy of the structure in kcal/mol </td></tr>
    <tr><td class="paramname">file</td><td>The file handle used to print to (print defaults to 'stdout' if(file == NULL) ) </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="gab69682373ccca1e0e28cc967eec07745"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vrna_file_connect </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>seq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>db</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>energy</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>identifier</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats_8h.html">ViennaRNA/file_formats.h</a>&gt;</code></p>

<p>Print a secondary structure as connect table. </p>
<p>Connect table file format looks like this: </p><pre class="fragment">300  ENERGY = 7.0  example
  1 G       0    2   22    1
  2 G       1    3   21    2
</pre><p> where the headerline is followed by 6 columns with:</p><ol type="1">
<li>Base number: index n</li>
<li>Base (A, C, G, T, U, X)</li>
<li>Index n-1 (0 if first nucleotide)</li>
<li>Index n+1 (0 if last nucleotide)</li>
<li>Number of the base to which n is paired. No pairing is indicated by 0 (zero).</li>
<li>Natural numbering.</li>
</ol>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">seq</td><td>The RNA sequence </td></tr>
    <tr><td class="paramname">db</td><td>The structure in dot-bracket format </td></tr>
    <tr><td class="paramname">energy</td><td>The free energy of the structure </td></tr>
    <tr><td class="paramname">identifier</td><td>An optional identifier for the sequence </td></tr>
    <tr><td class="paramname">file</td><td>The file handle used to print to (print defaults to 'stdout' if(file == NULL) ) </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="ga9b462e6f202594af5d3fa56e280d633f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vrna_file_bpseq </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>seq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>db</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats_8h.html">ViennaRNA/file_formats.h</a>&gt;</code></p>

<p>Print a secondary structure in bpseq format. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">seq</td><td>The RNA sequence </td></tr>
    <tr><td class="paramname">db</td><td>The structure in dot-bracket format </td></tr>
    <tr><td class="paramname">file</td><td>The file handle used to print to (print defaults to 'stdout' if(file == NULL) ) </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="ga31f4a6c2ea1495a6e4f9eb45a9f6193d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vrna_file_json </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>seq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>db</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>energy</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>identifier</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats_8h.html">ViennaRNA/file_formats.h</a>&gt;</code></p>

<p>Print a secondary structure in jsonformat. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">seq</td><td>The RNA sequence </td></tr>
    <tr><td class="paramname">db</td><td>The structure in dot-bracket format </td></tr>
    <tr><td class="paramname">energy</td><td>The free energy </td></tr>
    <tr><td class="paramname">identifier</td><td>An identifier for the sequence </td></tr>
    <tr><td class="paramname">file</td><td>The file handle used to print to (print defaults to 'stdout' if(file == NULL) ) </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="ga8cfb7e271efc9e1f34640acb85475639"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned int vrna_file_fasta_read_record </td>
          <td>(</td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>header</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>sequence</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char ***&#160;</td>
          <td class="paramname"><em>rest</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>options</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats_8h.html">ViennaRNA/file_formats.h</a>&gt;</code></p>

<p>Get a (fasta) data set from a file or stdin. </p>
<p>This function may be used to obtain complete datasets from a filehandle or stdin. A dataset is always defined to contain at least a sequence. If data starts with a fasta header, i.e. a line like </p><pre class="fragment">&gt;some header info </pre><p> then <a class="el" href="group__file__utils.html#ga8cfb7e271efc9e1f34640acb85475639" title="Get a (fasta) data set from a file or stdin. ">vrna_file_fasta_read_record()</a> will assume that the sequence that follows the header may span over several lines. To disable this behavior and to assign a single line to the argument 'sequence' one can pass <a class="el" href="group__utils.html#ga0de536599b881c787b0943a2671da476" title="Input switch for vrna_file_fasta_read_record(): &quot;never allow data to span more than one line&quot;...">VRNA_INPUT_NO_SPAN</a> in the 'options' argument. If no fasta header is read in the beginning of a data block, a sequence must not span over multiple lines!<br />
 Unless the options <a class="el" href="group__utils.html#ga0f6311f11bed1842e3a527ab27b294c6" title="Input switch for get_input_line(): &quot;do not skip comment lines&quot;. ">VRNA_INPUT_NOSKIP_COMMENTS</a> or <a class="el" href="group__utils.html#gab4db885222b3b69608310d7c7e63e286" title="Input switch for vrna_file_fasta_read_record(): &quot;do not skip empty lines&quot;. ">VRNA_INPUT_NOSKIP_BLANK_LINES</a> are passed, a sequence may be interrupted by lines starting with a comment character or empty lines.<br />
 A sequence is regarded as completely read if it was either assumed to not span over multiple lines, a secondary structure or structure constraint follows the sequence on the next line, or a new header marks the beginning of a new sequence...<br />
 All lines following the sequence (this includes comments) that do not initiate a new dataset according to the above definition are available through the line-array 'rest'. Here one can usually find the structure constraint or other information belonging to the current dataset. Filling of 'rest' may be prevented by passing <a class="el" href="group__utils.html#ga7a2e8c50a0c7ce82e60da1016e1367fd" title="Input switch for vrna_file_fasta_read_record(): &quot;do fill rest array&quot;. ">VRNA_INPUT_NO_REST</a> to the options argument.<br />
 </p><dl class="section note"><dt>Note</dt><dd>This function will exit any program with an error message if no sequence could be read! </dd>
<dd>
This function is NOT threadsafe! It uses a global variable to store information about the next data block.</dd></dl>
<p>The main purpose of this function is to be able to easily parse blocks of data in the header of a loop where all calculations for the appropriate data is done inside the loop. The loop may be then left on certain return values, e.g.: </p><div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno">    1</span>&#160;char *id, *seq, **rest;</div><div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;int  i;</div><div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;id = seq = NULL;</div><div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;rest = NULL;</div><div class="line"><a name="l00005"></a><span class="lineno">    5</span>&#160;while(!(vrna_file_fasta_read_record(&amp;id, &amp;seq, &amp;rest, NULL, 0) &amp; (VRNA_INPUT_ERROR | VRNA_INPUT_QUIT))){</div><div class="line"><a name="l00006"></a><span class="lineno">    6</span>&#160;  if(id) printf(&quot;%s\n&quot;, id);</div><div class="line"><a name="l00007"></a><span class="lineno">    7</span>&#160;  printf(&quot;%s\n&quot;, seq);</div><div class="line"><a name="l00008"></a><span class="lineno">    8</span>&#160;  if(rest)</div><div class="line"><a name="l00009"></a><span class="lineno">    9</span>&#160;    for(i=0;rest[i];i++){</div><div class="line"><a name="l00010"></a><span class="lineno">   10</span>&#160;      printf(&quot;%s\n&quot;, rest[i]);</div><div class="line"><a name="l00011"></a><span class="lineno">   11</span>&#160;      free(rest[i]);</div><div class="line"><a name="l00012"></a><span class="lineno">   12</span>&#160;    }</div><div class="line"><a name="l00013"></a><span class="lineno">   13</span>&#160;  free(rest);</div><div class="line"><a name="l00014"></a><span class="lineno">   14</span>&#160;  free(seq);</div><div class="line"><a name="l00015"></a><span class="lineno">   15</span>&#160;  free(id);</div><div class="line"><a name="l00016"></a><span class="lineno">   16</span>&#160;}</div></div><!-- fragment --><p> In the example above, the while loop will be terminated when <a class="el" href="group__file__utils.html#ga8cfb7e271efc9e1f34640acb85475639" title="Get a (fasta) data set from a file or stdin. ">vrna_file_fasta_read_record()</a> returns either an error, EOF, or a user initiated quit request.<br />
 As long as data is read from stdin (we are passing NULL as the file pointer), the id is printed if it is available for the current block of data. The sequence will be printed in any case and if some more lines belong to the current block of data each line will be printed as well.</p>
<dl class="section note"><dt>Note</dt><dd>Do not forget to free the memory occupied by header, sequence and rest!</dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">header</td><td>A pointer which will be set such that it points to the header of the record </td></tr>
    <tr><td class="paramname">sequence</td><td>A pointer which will be set such that it points to the sequence of the record </td></tr>
    <tr><td class="paramname">rest</td><td>A pointer which will be set such that it points to an array of lines which also belong to the record </td></tr>
    <tr><td class="paramname">file</td><td>A file handle to read from (if NULL, this function reads from stdin) </td></tr>
    <tr><td class="paramname">options</td><td>Some options which may be passed to alter the behavior of the function, use 0 for no options </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A flag with information about what the function actually did read </dd></dl>

</div>
</div>
<a class="anchor" id="gad37cbb63a05eed63ba25c91628409be0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char* vrna_extract_record_rest_structure </td>
          <td>(</td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>lines</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>length</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>option</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats_8h.html">ViennaRNA/file_formats.h</a>&gt;</code></p>

<p>Extract a dot-bracket structure string from (multiline)character array. </p>
<p>This function extracts a dot-bracket structure string from the 'rest' array as returned by <a class="el" href="group__file__utils.html#ga8cfb7e271efc9e1f34640acb85475639" title="Get a (fasta) data set from a file or stdin. ">vrna_file_fasta_read_record()</a> and returns it. All occurences of comments within the 'lines' array will be skipped as long as they do not break the structure string. If no structure could be read, this function returns NULL.</p>
<dl class="section pre"><dt>Precondition</dt><dd>The argument 'lines' has to be a 2-dimensional character array as obtained by <a class="el" href="group__file__utils.html#ga8cfb7e271efc9e1f34640acb85475639" title="Get a (fasta) data set from a file or stdin. ">vrna_file_fasta_read_record()</a> </dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__utils.html#ga8cfb7e271efc9e1f34640acb85475639" title="Get a (fasta) data set from a file or stdin. ">vrna_file_fasta_read_record()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">lines</td><td>The (multiline) character array to be parsed </td></tr>
    <tr><td class="paramname">length</td><td>The assumed length of the dot-bracket string (passing a value &lt; 1 results in no length limit) </td></tr>
    <tr><td class="paramname">option</td><td>Some options which may be passed to alter the behavior of the function, use 0 for no options </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The dot-bracket string read from lines or NULL </dd></dl>

</div>
</div>
<a class="anchor" id="ga646ebf45450a69a7f2533f9ecd283a32"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_file_SHAPE_read </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>file_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>length</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>default_value</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>sequence</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double *&#160;</td>
          <td class="paramname"><em>values</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats_8h.html">ViennaRNA/file_formats.h</a>&gt;</code></p>

<p>Read data from a given SHAPE reactivity input file. </p>
<p>This function parses the informations from a given file and stores the result in the preallocated string sequence and the double array values.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">file_name</td><td>Path to the constraints file </td></tr>
    <tr><td class="paramname">length</td><td>Length of the sequence (file entries exceeding this limit will cause an error) </td></tr>
    <tr><td class="paramname">default_value</td><td>Value for missing indices </td></tr>
    <tr><td class="paramname">sequence</td><td>Pointer to an array used for storing the sequence obtained from the SHAPE reactivity file </td></tr>
    <tr><td class="paramname">values</td><td>Pointer to an array used for storing the values obtained from the SHAPE reactivity file </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="gae33323c53765ecbbc410d9de2d495432"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="group__data__structures.html#ga8e4eb5e1bfc95776559575beb359af87">vrna_plist_t</a>* vrna_file_constraints_read </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>length</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>options</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats_8h.html">ViennaRNA/file_formats.h</a>&gt;</code></p>

<p>Read constraints from an input file. </p>
<p>This function reads constraint definitions from a file and converts them into an array of <a class="el" href="group__data__structures.html#ga8e4eb5e1bfc95776559575beb359af87" title="Typename for the base pair list repesenting data structure vrna_plist_s. ">vrna_plist_t</a> data structures. The data fields of each individual returned plist entry may adopt the following configurations:</p><ul>
<li>plist.i == plist.j <img class="formulaInl" alt="$ \rightarrow $" src="form_62.png"/> single nucleotide constraint</li>
<li>plist.i != plist.j <img class="formulaInl" alt="$ \rightarrow $" src="form_62.png"/> base pair constraint</li>
<li>plist.i == 0 <img class="formulaInl" alt="$ \rightarrow $" src="form_62.png"/> End of list </li>
</ul>

</div>
</div>
<a class="anchor" id="ga55a9ae6dfeecc1b3f0c2acf6fa796c15"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vrna_extract_record_rest_constraint </td>
          <td>(</td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>cstruc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>lines</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>option</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats_8h.html">ViennaRNA/file_formats.h</a>&gt;</code></p>

<p>Extract a hard constraint encoded as pseudo dot-bracket string. </p>
<dl class="deprecated"><dt><b><a class="el" href="deprecated.html#_deprecated000066">Deprecated:</a></b></dt><dd>Use <a class="el" href="group__file__utils.html#gad37cbb63a05eed63ba25c91628409be0" title="Extract a dot-bracket structure string from (multiline)character array. ">vrna_extract_record_rest_structure()</a> instead! </dd></dl>
<dl class="section pre"><dt>Precondition</dt><dd>The argument 'lines' has to be a 2-dimensional character array as obtained by <a class="el" href="group__file__utils.html#ga8cfb7e271efc9e1f34640acb85475639" title="Get a (fasta) data set from a file or stdin. ">vrna_file_fasta_read_record()</a> </dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__utils.html#ga8cfb7e271efc9e1f34640acb85475639" title="Get a (fasta) data set from a file or stdin. ">vrna_file_fasta_read_record()</a>, <a class="el" href="group__hard__constraints.html#ga13053547a2de5532b64b64d35e097ae1" title="Flag that is used to indicate the pipe &#39;|&#39; sign in pseudo dot-bracket notation of hard constraints...">VRNA_CONSTRAINT_DB_PIPE</a>, <a class="el" href="group__hard__constraints.html#ga369bea82eae75fbe626f409fa425747e" title="dot &#39;.&#39; switch for structure constraints (no constraint at all) ">VRNA_CONSTRAINT_DB_DOT</a>, <a class="el" href="group__hard__constraints.html#ga7283bbe0f8954f7b030ecc3f2d1932b2" title="&#39;x&#39; switch for structure constraint (base must not pair) ">VRNA_CONSTRAINT_DB_X</a> <a class="el" href="constraints__hard_8h.html#ad54c1315a47d55653dcaa5de6e544b77" title="angle brackets &#39;&lt;&#39;, &#39;&gt;&#39; switch for structure constraint (paired downstream/upstream) ...">VRNA_CONSTRAINT_DB_ANG_BRACK</a>, <a class="el" href="group__hard__constraints.html#gac17b034852c914bc5879954c65d7e74b" title="round brackets &#39;(&#39;,&#39;)&#39; switch for structure constraint (base i pairs base j) ">VRNA_CONSTRAINT_DB_RND_BRACK</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">cstruc</td><td>A pointer to a character array that is used as pseudo dot-bracket output </td></tr>
    <tr><td class="paramname">lines</td><td>A 2-dimensional character array with the extension lines from the FASTA input </td></tr>
    <tr><td class="paramname">option</td><td>The option flags that define the behavior and recognition pattern of this function </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="gafd194a69af9d92b5b0412a7627ac1595"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned int read_record </td>
          <td>(</td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>header</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>sequence</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char ***&#160;</td>
          <td class="paramname"><em>rest</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>options</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats_8h.html">ViennaRNA/file_formats.h</a>&gt;</code></p>

<p>Get a data record from stdin. </p>
<dl class="deprecated"><dt><b><a class="el" href="deprecated.html#_deprecated000068">Deprecated:</a></b></dt><dd>This function is deprecated! Use <a class="el" href="group__file__utils.html#ga8cfb7e271efc9e1f34640acb85475639" title="Get a (fasta) data set from a file or stdin. ">vrna_file_fasta_read_record()</a> as a replacment.</dd></dl>

</div>
</div>
<a class="anchor" id="gad02d5d12bda54611c915a1019323b7be"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_file_msa_read </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char ***&#160;</td>
          <td class="paramname"><em>names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char ***&#160;</td>
          <td class="paramname"><em>aln</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>options</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats__msa_8h.html">ViennaRNA/file_formats_msa.h</a>&gt;</code></p>

<p>Read a multiple sequence alignment from file. </p>
<p>This function reads the (first) multiple sequence alignment from an input file. The read alignment is split into the sequence id/name part and the actual sequence information and stored in memory as arrays of ids/names and sequences. If the alignment file format allows for additional information, such as an ID of the entire alignment or consensus structure information, this data is retrieved as well and made available. The <code>options</code> parameter allows to specify the set of alignment file formats that should be used to retrieve the data. If 0 is passed as option, the list of alignment file formats defaults to <a class="el" href="group__file__utils.html#ga4acc255373831856a8417b68de8a94c0" title="Option flag indicating the set of default file formats. ">VRNA_FILE_FORMAT_MSA_DEFAULT</a>.</p>
<p>Currently, the list of parsable multiple sequence alignment file formats consists of:</p><ul>
<li><a class="el" href="file_formats.html#msa-formats-clustal">ClustalW format</a></li>
<li><a class="el" href="file_formats.html#msa-formats-stockholm">Stockholm 1.0 format</a></li>
<li><a class="el" href="file_formats.html#msa-formats-fasta">FASTA (Pearson) format</a></li>
<li><a class="el" href="file_formats.html#msa-formats-maf">MAF format</a></li>
</ul>
<dl class="section note"><dt>Note</dt><dd>After successfully reading an alignment, this function performs a validation of the data that includes uniqueness of the sequence identifiers, and equal sequence lengths. This check can be deactivated by passing <a class="el" href="group__file__utils.html#ga229fb3778ecabea4782902b69fa48fd1" title="Option flag to disable validation of the alignment. ">VRNA_FILE_FORMAT_MSA_NOCHECK</a> in the <code>options</code> parameter.</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__utils.html#gad4203a438622b2df7bc2f16578d54799" title="Read a multiple sequence alignment from file handle. ">vrna_file_msa_read_record()</a>, <a class="el" href="group__file__utils.html#ga79a23de2c7249f2cccd762e475c81859" title="Option flag indicating ClustalW formatted files. ">VRNA_FILE_FORMAT_MSA_CLUSTAL</a>, <a class="el" href="group__file__utils.html#ga62be992445cd8ab2ad7a8fded944338b" title="Option flag indicating Stockholm 1.0 formatted files. ">VRNA_FILE_FORMAT_MSA_STOCKHOLM</a>, <a class="el" href="group__file__utils.html#gacf6274a2c825f34e4131404665b00604" title="Option flag indicating FASTA (Pearson) formatted files. ">VRNA_FILE_FORMAT_MSA_FASTA</a>, <a class="el" href="group__file__utils.html#gab3186d774c01570d2a47d0308eec4927" title="Option flag indicating MAF formatted files. ">VRNA_FILE_FORMAT_MSA_MAF</a>, <a class="el" href="group__file__utils.html#ga4acc255373831856a8417b68de8a94c0" title="Option flag indicating the set of default file formats. ">VRNA_FILE_FORMAT_MSA_DEFAULT</a>, <a class="el" href="group__file__utils.html#ga229fb3778ecabea4782902b69fa48fd1" title="Option flag to disable validation of the alignment. ">VRNA_FILE_FORMAT_MSA_NOCHECK</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>The name of input file that contains the alignment </td></tr>
    <tr><td class="paramname">names</td><td>An address to the pointer where sequence identifiers should be written to </td></tr>
    <tr><td class="paramname">aln</td><td>An address to the pointer where aligned sequences should be written to </td></tr>
    <tr><td class="paramname">id</td><td>An address to the pointer where the alignment ID should be written to (Maybe NULL) </td></tr>
    <tr><td class="paramname">structure</td><td>An address to the pointer where consensus structure information should be written to (Maybe NULL) </td></tr>
    <tr><td class="paramname">options</td><td>Options to manipulate the behavior of this function </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The number of sequences in the alignment, or -1 if no alignment record could be found </dd></dl>

</div>
</div>
<a class="anchor" id="gad4203a438622b2df7bc2f16578d54799"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_file_msa_read_record </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>fp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char ***&#160;</td>
          <td class="paramname"><em>names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char ***&#160;</td>
          <td class="paramname"><em>aln</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>options</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats__msa_8h.html">ViennaRNA/file_formats_msa.h</a>&gt;</code></p>

<p>Read a multiple sequence alignment from file handle. </p>
<p>Similar to <a class="el" href="group__file__utils.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file. ">vrna_file_msa_read()</a>, this function reads a multiple sequence alignment from an input file handle. Since using a file handle, this function is not limited to the first alignment record, but allows for looping over all alignments within the input.</p>
<p>The read alignment is split into the sequence id/name part and the actual sequence information and stored in memory as arrays of ids/names and sequences. If the alignment file format allows for additional information, such as an ID of the entire alignment or consensus structure information, this data is retrieved as well and made available. The <code>options</code> parameter allows to specify the alignment file format used to retrieve the data. A single format must be specified here, see <a class="el" href="group__file__utils.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file. ">vrna_file_msa_detect_format()</a> for helping to determine the correct MSA file format.</p>
<p>Currently, the list of parsable multiple sequence alignment file formats consists of:</p><ul>
<li><a class="el" href="file_formats.html#msa-formats-clustal">ClustalW format</a></li>
<li><a class="el" href="file_formats.html#msa-formats-stockholm">Stockholm 1.0 format</a></li>
<li><a class="el" href="file_formats.html#msa-formats-fasta">FASTA (Pearson) format</a></li>
<li><a class="el" href="file_formats.html#msa-formats-maf">MAF format</a></li>
</ul>
<dl class="section note"><dt>Note</dt><dd>After successfully reading an alignment, this function performs a validation of the data that includes uniqueness of the sequence identifiers, and equal sequence lengths. This check can be deactivated by passing <a class="el" href="group__file__utils.html#ga229fb3778ecabea4782902b69fa48fd1" title="Option flag to disable validation of the alignment. ">VRNA_FILE_FORMAT_MSA_NOCHECK</a> in the <code>options</code> parameter.</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__utils.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file. ">vrna_file_msa_read()</a>, <a class="el" href="group__file__utils.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file. ">vrna_file_msa_detect_format()</a>, <a class="el" href="group__file__utils.html#ga79a23de2c7249f2cccd762e475c81859" title="Option flag indicating ClustalW formatted files. ">VRNA_FILE_FORMAT_MSA_CLUSTAL</a>, <a class="el" href="group__file__utils.html#ga62be992445cd8ab2ad7a8fded944338b" title="Option flag indicating Stockholm 1.0 formatted files. ">VRNA_FILE_FORMAT_MSA_STOCKHOLM</a>, <a class="el" href="group__file__utils.html#gacf6274a2c825f34e4131404665b00604" title="Option flag indicating FASTA (Pearson) formatted files. ">VRNA_FILE_FORMAT_MSA_FASTA</a>, <a class="el" href="group__file__utils.html#gab3186d774c01570d2a47d0308eec4927" title="Option flag indicating MAF formatted files. ">VRNA_FILE_FORMAT_MSA_MAF</a>, <a class="el" href="group__file__utils.html#ga4acc255373831856a8417b68de8a94c0" title="Option flag indicating the set of default file formats. ">VRNA_FILE_FORMAT_MSA_DEFAULT</a>, <a class="el" href="group__file__utils.html#ga229fb3778ecabea4782902b69fa48fd1" title="Option flag to disable validation of the alignment. ">VRNA_FILE_FORMAT_MSA_NOCHECK</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">fp</td><td>The file pointer the data will be retrieved from </td></tr>
    <tr><td class="paramname">names</td><td>An address to the pointer where sequence identifiers should be written to </td></tr>
    <tr><td class="paramname">aln</td><td>An address to the pointer where aligned sequences should be written to </td></tr>
    <tr><td class="paramname">id</td><td>An address to the pointer where the alignment ID should be written to (Maybe NULL) </td></tr>
    <tr><td class="paramname">structure</td><td>An address to the pointer where consensus structure information should be written to (Maybe NULL) </td></tr>
    <tr><td class="paramname">options</td><td>Options to manipulate the behavior of this function </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The number of sequences in the alignment, or -1 if no alignment record could be found </dd></dl>

</div>
</div>
<a class="anchor" id="gade4fa8136ebb2d0f7eb3f8b59a8658e3"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned int vrna_file_msa_detect_format </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>options</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="file__formats__msa_8h.html">ViennaRNA/file_formats_msa.h</a>&gt;</code></p>

<p>Detect the format of a multiple sequence alignment file. </p>
<p>This function attempts to determine the format of a file that supposedly contains a multiple sequence alignment (MSA). This is useful in cases where a MSA file contains more than a single record and therefore <a class="el" href="group__file__utils.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file. ">vrna_file_msa_read()</a> can not be applied, since it only retrieves the first. Here, one can try to guess the correct file format using this function and then loop over the file, record by record using one of the low-level record retrieval functions for the corresponding MSA file format.</p>
<dl class="section note"><dt>Note</dt><dd>This function parses the entire first record within the specified file. As a result, it returns <a class="el" href="group__file__utils.html#gabdc948f547e550125de3e7c65878400c" title="Return flag of vrna_file_msa_detect_format() to indicate unknown or malformatted alignment. ">VRNA_FILE_FORMAT_MSA_UNKNOWN</a> not only if it can't detect the file's format, but also in cases where the file doesn't contain sequences!</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__utils.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file. ">vrna_file_msa_read()</a>, vrna_file_stockholm_read_record(), vrna_file_clustal_read_record(), <a class="el" href="group__file__utils.html#ga8cfb7e271efc9e1f34640acb85475639" title="Get a (fasta) data set from a file or stdin. ">vrna_file_fasta_read_record()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>The name of input file that contains the alignment </td></tr>
    <tr><td class="paramname">options</td><td>Options to manipulate the behavior of this function </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The MSA file format, or <a class="el" href="group__file__utils.html#gabdc948f547e550125de3e7c65878400c" title="Return flag of vrna_file_msa_detect_format() to indicate unknown or malformatted alignment. ">VRNA_FILE_FORMAT_MSA_UNKNOWN</a> </dd></dl>

</div>
</div>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated on Tue Sep 6 2016 15:07:12 for RNAlib-2.2.10 by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.11 </li>
  </ul>
</div>
</body>
</html>
